<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Statistical Inference for Biology: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Statistical Inference for Biology
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Statistical Inference for Biology
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Statistical Inference for Biology
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="inference-rv-dists.html">2. Inference</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="inference-pse.html">3. Populations, Samples and Estimates</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="inference-clt.html">4. Central Limit Theorem and the t-distribution</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="inference-clt-practice.html">5. Central Limit Theorem in practice</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="inference-ttests-practice.html">6. t-tests in practice</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="inference-ci.html">7. Confidence Intervals</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="inference-power-calc.html">8. Power Calculations</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="inference-montecarlo.html">9. Monte Carlo simulation</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="inference-permutations.html">10. Permutations</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="inference-assoc-tests.html">11. Association tests</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush13">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading13">
        <a href="eda-intro.html">12. Exploratory Data Analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush14">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading14">
        <a href="eda-avoid-plots.html">13. Plots to avoid</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html">Glossary</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Introduction</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/introduction.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is statistical inference?</li>
<li>Why do biomedical researchers need to learn statistics now?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how technology has changed biomedical measurements from past
to present.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>Technological changes in biomedical research drive data production in
greater quantity and complexity. High-throughput technologies, such as
sequencing technologies, produce data whose size and complexity require
sophisticated statistical skills to avoid being fooled by patterns
arising by chance. In the past researchers would measure, for example,
the transcription levels of a single gene of interest. Now it is
possible to measure all genes at once, often 20,000 or more depending on
the organism. Technological advances like these have driven a change
from hypothesis to discovery-driven research. This means that statistics
and data analysis in the life sciences are more important than ever.</p>
<p>This lesson will introduce the statistical concepts and data analysis
skills needed for success in data-driven life science research. We start
with one of the most important topics in statistics and in the life
sciences: statistical inference. Inference is the use of probability to
learn population characteristics from data. A typical example is
determining if two groups (for example, cases versus controls) are
different on average. Specific topics include:</p>
<ul>
<li>p-values</li>
<li>the t-test</li>
<li>confidence intervals</li>
<li>association tests</li>
<li>permutation tests</li>
<li>and statistical power</li>
</ul>
<p>We make use of approximations made possible by mathematical theory,
such as the Central Limit Theorem, as well as techniques made possible
by modern computing.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Novel technologies produce data in great complexity and scale,
requiring more sophisticated understanding of statistics.</li>
</ul>
</div>
</div>
</div></section><section id="aio-inference-rv-dists"><p>Content from <a href="inference-rv-dists.html">Inference</a></p>
<hr>
<p>Last updated on 2025-08-17 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-rv-dists.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What does inference mean?</li>
<li>Why do we need p-values and confidence intervals?</li>
<li>What is a random variable?</li>
<li>What exactly is a distribution?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe the statistical concepts underlying p-values and confidence
intervals.</li>
<li>Explain random variables and null distributions using R
programming.</li>
<li>Compute p-values and confidence intervals using R programming.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>This section introduces the statistical concepts necessary to
understand p-values and confidence intervals. These terms are ubiquitous
in the life science literature. Let’s use <a href="https://diabetes.diabetesjournals.org/content/53/suppl_3/S215.full" class="external-link">this
paper</a> as an example.</p>
<p>Note that the abstract has this statement:</p>
<blockquote>
<p>“Body weight was higher in mice fed the high-fat diet already after
the first week, due to higher dietary intake in combination with lower
metabolic efficiency.”</p>
</blockquote>
<p>To support this claim they provide the following in the results
section:</p>
<blockquote>
<p>“Already during the first week after introduction of high-fat diet,
body weight increased significantly more in the high-fat diet-fed mice
(+ 1.6 ± 0.1 g) than in the normal diet-fed mice (+ 0.2 ± 0.1 g; P &lt;
0.001).”</p>
</blockquote>
<p>What does P &lt; 0.001 mean? Why are the ± included? We will learn
what this means and learn to compute these values in R. The first step
is to understand random variables. To do this, we will use data from a
mouse database (provided by Karen Svenson via Gary Churchill and Dan
Gatti and partially funded by P50 GM070683). We will import the data
into R and explain random variables and null distributions using R
programming. See the <a href="%22https://smcclatchy.github.io/statistical-inference-for-biology/setup/%22" class="external-link">Setup</a>
instructions to import the data.</p>
<div class="section level5">
<h5 id="our-first-look-at-data">Our first look at data<a class="anchor" aria-label="anchor" href="#our-first-look-at-data"></a>
</h5>
<p>We are interested in determining if following a given diet makes mice
heavier after several weeks. This data was produced by ordering 24 mice
from The Jackson Lab and randomly assigning either chow or high fat (hf)
diet. After several weeks, the scientists weighed each mouse and
obtained this data (<code>head</code> just shows us the first 6
rows):</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fWeights</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"./data/femaleMiceWeights.csv"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">fWeights</span><span class="op">)</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Diet Bodyweight
1 chow      21.51
2 chow      28.14
3 chow      24.04
4 chow      23.45
5 chow      23.68
6 chow      19.79</code></pre>
</div>
<p>If you would like to view the entire data set with RStudio:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">fWeights</span><span class="op">)</span> </span></code></pre>
</div>
<p>So are the hf mice heavier? Mouse 24 at 20.73 grams is one of the
lightest mice, while Mouse 21 at 34.02 grams is one of the heaviest.
Both are on the hf diet. Just from looking at the data, we see there is
<em>variability</em>. Claims such as the previous (that body weight
increased significantly in high-fat diet-fed mice) usually refer to the
averages. So let’s look at the average of each group:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">unlist</span></span>
<span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">unlist</span></span>
<span><span class="fu">print</span><span class="op">(</span> <span class="fu">mean</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 26.83417</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">print</span><span class="op">(</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23.81333</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obsdiff</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">obsdiff</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.020833</code></pre>
</div>
<p>So the hf diet mice are about 10% heavier. Are we done? Why do we
need p-values and confidence intervals? The reason is that these
averages are random variables. They can take many values.</p>
<p>If we repeat the experiment, we obtain 24 new mice from The Jackson
Laboratory and, after randomly assigning them to each diet, we get a
different mean. Every time we repeat this experiment, we get a different
value. We call this type of quantity a <em>random variable</em>.</p>
</div>
</div>
<div class="section level3">
<h3 id="random-variables">Random Variables<a class="anchor" aria-label="anchor" href="#random-variables"></a>
</h3>
<p>Let’s explore random variables further. Imagine that we actually have
the weight of all control female mice and can upload them to R. In
Statistics, we refer to this as <em>the population</em>. These are all
the control mice available from which we sampled 24. Note that in
practice we do not have access to the population. We have a special
dataset that we are using here to illustrate concepts.</p>
<p>Now let’s sample 12 mice three times and see how the average
changes.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">population</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleControlsPopulation.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span><span class="op">$</span><span class="va">Bodyweight</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23.47667</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span><span class="op">$</span><span class="va">Bodyweight</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 25.12583</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span><span class="op">$</span><span class="va">Bodyweight</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23.72333</code></pre>
</div>
<p>Note how the average varies. We can continue to do this repeatedly
and start learning something about the distribution of this random
variable.</p>
</div>
<div class="section level3">
<h3 id="the-null-hypothesis">The Null Hypothesis<a class="anchor" aria-label="anchor" href="#the-null-hypothesis"></a>
</h3>
<p>Now let’s go back to our average difference of <code>obsdiff</code>.
As scientists we need to be skeptics. How do we know that this
<code>obsdiff</code> is due to the diet? What happens if we give all 24
mice the same diet? Will we see a difference this big? Statisticians
refer to this scenario as the <em>null hypothesis</em>. The name “null”
is used to remind us that we are acting as skeptics: we give credence to
the possibility that there is no difference.</p>
<p>Because we have access to the population, we can actually observe as
many values as we want of the difference of the averages when the diet
has no effect. We can do this by randomly sampling 24 control mice,
giving them the same diet, and then recording the difference in mean
between two randomly split groups of 12 and 12. Here is this process
written in R code:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 12 control mice</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span><span class="op">$</span><span class="va">Bodyweight</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="co">## another 12 control mice that we act as if they were not</span></span>
<span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span><span class="op">$</span><span class="va">Bodyweight</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] -0.2391667</code></pre>
</div>
<p>Now let’s do it 10,000 times. We will use a “for-loop”, an operation
that lets us automate this (a simpler approach that, we will learn
later, is to use <code>replicate</code>).</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">null</span> <span class="op">&lt;-</span> <span class="fu">vector</span><span class="op">(</span><span class="st">"numeric"</span>, <span class="va">n</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span><span class="op">$</span><span class="va">Bodyweight</span>, <span class="fl">12</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span><span class="op">$</span><span class="va">Bodyweight</span>, <span class="fl">12</span><span class="op">)</span></span>
<span>  <span class="va">null</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>The values in <code>null</code> form what we call the <em>null
distribution</em>. We will define this more formally below. By the way,
the loop above is a <em>Monte Carlo</em> simulation to obtain 10,000
outcomes of the random variable under the null hypothesis. Simulations
can be used to check theoretical or analytical results. For more
information about Monte Carlo simulations, visit <a href="https://leanpub.com/dataanalysisforthelifesciences" class="external-link">Data Analysis
for the Life Sciences</a>.</p>
<p>So what percent of the 10,000 are bigger than
<code>obsdiff</code>?</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mean</span><span class="op">(</span><span class="va">null</span> <span class="op">&gt;=</span> <span class="va">obsdiff</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.0123</code></pre>
</div>
<p>Only a small percent of the 10,000 simulations. As skeptics what do
we conclude? When there is no diet effect, we see a difference as big as
the one we observed only 1.5% of the time. This is what is known as a
p-value, which we will define more formally later in the book.</p>
</div>
<div class="section level3">
<h3 id="distributions">Distributions<a class="anchor" aria-label="anchor" href="#distributions"></a>
</h3>
<p>We have explained what we mean by <em>null</em> in the context of
null hypothesis, but what exactly is a distribution?</p>
<p>The simplest way to think of a <em>distribution</em> is as a compact
description of many numbers. For example, suppose you have measured the
heights of all men in a population. Imagine you need to describe these
numbers to someone that has no idea what these heights are, such as an
alien that has never visited Earth. Suppose all these heights are
contained in the following dataset:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">father.son</span> <span class="op">&lt;-</span> <span class="fu">UsingR</span><span class="fu">::</span><span class="va">father.son</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">father.son</span><span class="op">$</span><span class="va">fheight</span></span></code></pre>
</div>
<p>One approach to summarizing these numbers is to simply list them all
out for the alien to see. Here are 10 randomly selected heights of
1,078:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">x</span>, <span class="fl">10</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 67.7 72.5 64.7 62.7 66.1 67.5 68.2 61.8 69.7 71.2</code></pre>
</div>
<div class="section level5">
<h5 id="cumulative-distribution-function">Cumulative Distribution Function<a class="anchor" aria-label="anchor" href="#cumulative-distribution-function"></a>
</h5>
<p>Scanning through these numbers, we start to get a rough idea of what
the entire list looks like, but it is certainly inefficient. We can
quickly improve on this approach by defining and visualizing a
<em>distribution</em>. To define a distribution we compute, for all
possible values of <i>a</i>, the proportion of numbers in our list that
are below <i>a</i>. We use the following notation:</p>
<figure><img src="fig/02-cum-dist-function.png" class="figure mx-auto d-block"></figure><p>This is called the cumulative distribution function (CDF). When the
CDF is derived from data, as opposed to theoretically, we also call it
the empirical CDF (ECDF). The ECDF for the height data looks like
this:</p>
<figure style="text-align: center"><img src="fig/inference-rv-dists-rendered-ecdf-1.png" alt="Empirical cumulative distribution function for height." class="figure mx-auto d-block"><figcaption>
Empirical cumulative distribution function for height.
</figcaption></figure>
</div>
<div class="section level5">
<h5 id="histograms">Histograms<a class="anchor" aria-label="anchor" href="#histograms"></a>
</h5>
<p>Although the empirical CDF concept is widely discussed in statistics
textbooks, the plot is actually not very popular in practice. The reason
is that histograms give us the same information and are easier to
interpret. Histograms show us the proportion of values in intervals:</p>
<p><img src="fig/02-proba-lessthan-b.png" class="figure"> Plotting these heights as
bars is what we call a <em>histogram</em>. It is a more useful plot
because we are usually more interested in intervals, such and such
percent are between 70 inches and 71 inches, etc., rather than the
percent less than a particular height. Here is a histogram of
heights:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">x</span>, xlab<span class="op">=</span><span class="st">"Height (in inches)"</span>, main<span class="op">=</span><span class="st">"Adult men heights"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-rv-dists-rendered-histogram-1.png" alt="Histogram for heights." class="figure mx-auto d-block"><figcaption>
Histogram for heights.
</figcaption></figure><p>Showing this plot to the alien is much more informative than showing
numbers. With this simple plot, we can approximate the number of
individuals in any given interval. For example, there are about 70
individuals over six feet (72 inches) tall.</p>
</div>
</div>
<div class="section level3">
<h3 id="probability-distribution">Probability Distribution<a class="anchor" aria-label="anchor" href="#probability-distribution"></a>
</h3>
<p>Summarizing lists of numbers is one powerful use of a distribution.
An even more important use is describing the possible outcomes of a
random variable. Unlike a fixed list of numbers, we don’t actually
observe all possible outcomes of random variables, so instead of
describing proportions, we describe probabilities. For instance, if we
pick a random height from our list, then the probability of it falling
between <i>a</i> and <i>b</i> is denoted with:</p>
<p><img src="fig/02-proba-lt-random.png" class="figure"> Note that the <i>X</i> is
now capitalized to distinguish it as a random variable and that the
equation above defines the probability distribution of the random
variable. Knowing this distribution is incredibly useful in science. For
example, in the case above, if we know the distribution of the
difference in mean of mouse weights when the null hypothesis is true,
referred to as the <em>null distribution</em>, we can compute the
probability of observing a value as large as we did, referred to as a
<em>p-value</em>. In a previous section we ran what is called a
<em>Monte Carlo</em> simulation (we will provide more details on Monte
Carlo simulation in a later section) and we obtained 10,000 outcomes of
the random variable under the null hypothesis.</p>
<p>The observed values will amount to a histogram. From a histogram of
the <code>null</code> vector we calculated earlier, we can see that
values as large as <code>obsdiff</code> are relatively rare:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">null</span>, freq<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="va">obsdiff</span>, col<span class="op">=</span><span class="st">"red"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-rv-dists-rendered-null_and_obs-1.png" alt="Null distribution with observed difference marked with vertical red line." class="figure mx-auto d-block"><figcaption>
Null distribution with observed difference marked with vertical red
line.
</figcaption></figure><p>An important point to keep in mind here is that while we defined
Pr<i>(a)</i> by counting cases, we will learn that, in some
circumstances, mathematics gives us formulas for Pr<i>(a)</i> that save
us the trouble of computing them as we did here. One example of this
powerful approach uses the normal distribution approximation.</p>
</div>
<div class="section level3">
<h3 id="normal-distribution">Normal Distribution<a class="anchor" aria-label="anchor" href="#normal-distribution"></a>
</h3>
<p>The probability distribution we see above approximates one that is
very common in nature: the bell curve, also known as the normal
distribution or Gaussian distribution. When the histogram of a list of
numbers approximates the normal distribution, we can use a convenient
mathematical formula to approximate the proportion of values or outcomes
in any given interval:</p>
<figure><img src="fig/02-normal-dist-hist.png" class="figure mx-auto d-block"></figure><p>While the formula may look intimidating, don’t worry, you will never
actually have to type it out, as it is stored in a more convenient form
(as <code>pnorm</code> in R which sets <em>a</em> to -∞, and takes
<em>b</em> as an argument).</p>
<p>Here μ and σ are referred to as the mean and the standard deviation
of the population (we explain these in more detail in another section).
If this <em>normal approximation</em> holds for our list, then the
population mean and variance of our list can be used in the formula
above. An example of this would be when we noted above that only 1.5% of
values on the null distribution were above <code>obsdiff</code>. We can
compute the proportion of values below a value <code>x</code> with
<code>pnorm(x, mu, sigma)</code> without knowing all the values. The
normal approximation works very well here:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu">pnorm</span><span class="op">(</span><span class="va">obsdiff</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">null</span><span class="op">)</span>, <span class="fu">sd</span><span class="op">(</span><span class="va">null</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.01311009</code></pre>
</div>
<p>Later, we will learn that there is a mathematical explanation for
this. A very useful characteristic of this approximation is that one
only needs to know μ and σ to describe the entire distribution. From
this, we can compute the proportion of values in any interval.</p>
<p>Refer to the histogram of null values above. The code we just ran
represents everything to the right of the vertical red line, or 1 minus
everything to the left. Try running this code without the
<code>1 -</code> to understand this better.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pnorm</span><span class="op">(</span><span class="va">obsdiff</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">null</span><span class="op">)</span>, <span class="fu">sd</span><span class="op">(</span><span class="va">null</span><span class="op">)</span><span class="op">)</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.9868899</code></pre>
</div>
<p>This value represents everything to the left of the vertical red line
in the null histogram above.</p>
<div class="section level5">
<h5 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h5>
<p>So computing a p-value for the difference in diet for the mice was
pretty easy, right? But why are we not done? To make the calculation, we
did the equivalent of buying all the mice available from The Jackson
Laboratory and performing our experiment repeatedly to define the null
distribution. Yet this is not something we can do in practice.
Statistical Inference is the mathematical theory that permits you to
approximate this with only the data from your sample, i.e. the original
24 mice. We will focus on this in the following sections.</p>
</div>
<div class="section level5">
<h5 id="setting-the-random-seed">Setting the random seed<a class="anchor" aria-label="anchor" href="#setting-the-random-seed"></a>
</h5>
<p>Before we continue, we briefly explain the following important line
of code:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> </span></code></pre>
</div>
<p>Throughout this lesson, we use random number generators. This implies
that many of the results presented can actually change by chance,
including the correct answer to problems. One way to ensure that results
do not change is by setting R’s random number generation seed. For more
on the topic please read the help file:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">set.seed</span></span></code></pre>
</div>
<p>Even better is to review an example. R has a built-in vector of
characters called <code>LETTERS</code> that contains upper case letters
of the Roman alphabet. We can take a sample of 5 letters with the
following code, which when repeated will give a different set of 5
letters each times.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LETTERS</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S"
[20] "T" "U" "V" "W" "X" "Y" "Z"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Y" "D" "G" "A" "B"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "W" "K" "N" "R" "S"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sample</span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A" "U" "Y" "J" "V"</code></pre>
</div>
<p>If we set a seed, we will get the same sample of letters each
time.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Y" "D" "G" "A" "B"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Y" "D" "G" "A" "B"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Y" "D" "G" "A" "B"</code></pre>
</div>
<p>When we set a seed we ensure that we get the same results from random
number generation, which is used in sampling with
<code>sample</code>.</p>
<p>For the following exercises, we will be using the female controls
population dataset that we read into a variable called
<code>population</code>. Here <code>population</code> represents the
weights for the entire population of female mice. To remind ourselves
about this data set, run the following:</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'data.frame':	225 obs. of  1 variable:
 $ Bodyweight: num  27 24.8 27 28.1 23.6 ...</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Bodyweight
1      27.03
2      24.80
3      27.02
4      28.07
5      23.55
6      22.72</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">population</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   Bodyweight
 Min.   :15.51
 1st Qu.:21.51
 Median :23.54
 Mean   :23.89
 3rd Qu.:26.08
 Max.   :36.84  </code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>What is the average of these weights?</li>
<li>After setting the seed at 1, (<code>set.seed(1)</code>) take a
random sample of size 5.<br>
What is the absolute value (<code>abs()</code>) of the difference
between the average of the sample and the average of all the
values?</li>
<li>After setting the seed at 5, <code>set.seed(5)</code> take a random
sample of size 5. What is the absolute value of the difference between
the average of the sample and the average of all the values?</li>
<li>Why are the answers from 2 and 3 different?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Because we made a coding mistake.<br>
</li>
<li>Because the average of the population weights is random.<br>
</li>
<li>Because the average of the samples is a random variable.<br>
</li>
<li>All of the above.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Solution to Exercise 1
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li><p><code>mean(population$Bodyweight)</code></p></li>
<li><p><code>set.seed(1)</code><br><code>meanOfSample1 &lt;- mean(sample(population$Bodyweight, 5))</code><br><code>abs(meanOfSample1 - mean(population$Bodyweight))</code></p></li>
<li><p><code>set.seed(5)</code><br><code>meanOfSample2 &lt;- mean(sample(population$Bodyweight, 5))</code><br><code>abs(meanOfSample2 - mean(population$Bodyweight))</code></p></li>
<li><ol start="3" style="list-style-type: upper-alpha">
<li>Because the average of the samples is a random variable.</li>
</ol></li>
</ol>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Set the seed at 1, then using a for-loop take a random sample of 5
mice 1,000 times. Save these averages. What percent of these 1,000
averages are more than 1 gram away from the average of the
population?</li>
<li>We are now going to increase the number of times we redo the sample
from 1,000 to 10,000. Set the seed at 1, then using a for-loop take a
random sample of 5 mice 10,000 times. Save these averages. What percent
of these 10,000 averages are more than 1 gram away from the average of
the population?</li>
<li>Note that the answers to the previous two questions barely changed.
This is expected. The way we think about the random value distributions
is as the distribution of the list of values obtained if we repeated the
experiment an infinite number of times. On a computer, we can’t perform
an infinite number of iterations so instead, for our examples, we
consider 1,000 to be large enough, thus 10,000 is as well. Now if
instead we change the sample size, then we change the random variable
and thus its distribution.<br>
Set the seed at 1, then using a for-loop take a random sample of 50 mice
1,000 times. Save these averages. What percent of these 1,000 averages
are more than 1 gram away from the average of the population?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Solution to Exercise 2
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>
<code>set.seed(1)</code><br><code>n &lt;- 1000</code><br><code>meanSampleOf5 &lt;- vector("numeric", n)</code><br><code>for (i in 1:n) {</code><br><code>meanSampleOf5[i] &lt;- mean(sample(population$Bodyweight, 5)</code><br><code>}</code>meanSampleOf5<code></code>mean(population$Bodyweight)`</li>
</ol>
<div class="section level2">
<h2 id="histogram-of-sample-means">histogram of sample means<a class="anchor" aria-label="anchor" href="#histogram-of-sample-means"></a>
</h2>
<p><code>hist(meanSampleof5)</code></p>
</div>
<div class="section level2">
<h2 id="mean-population-weight-plus-one-gram-and-minus-one-gram">mean population weight plus one gram and minus one gram<a class="anchor" aria-label="anchor" href="#mean-population-weight-plus-one-gram-and-minus-one-gram"></a>
</h2>
<p><code>meanPopWeight &lt;- mean(population$Bodyweight)</code> # mean
<code>sdPopWeight &lt;- sd(population$Bodyweight)</code> # standard
deviation <code>abline(v = meanPopWeight, col = "blue", lwd = 2)</code>
<code>abline(v = meanPopWeight + 1, col = "red", lwd = 2)</code>
<code>abline(v = meanPopWeight - 1, col = "red", lwd = 2)</code></p>
</div>
<div class="section level2">
<h2 id="proportion-below-mean-population-weight-minus-1-gram">proportion below mean population weight minus 1 gram<a class="anchor" aria-label="anchor" href="#proportion-below-mean-population-weight-minus-1-gram"></a>
</h2>
<p><code>pnorm(meanPopWeight - 1, mean = meanPopWeight, sd = sdPopWeight)</code></p>
</div>
<div class="section level2">
<h2 id="proportion-greater-than-mean-population-weight-plus-1-gram">proportion greater than mean population weight plus 1 gram<a class="anchor" aria-label="anchor" href="#proportion-greater-than-mean-population-weight-plus-1-gram"></a>
</h2>
<p><code>1 - pnorm(meanPopWeight + 1, mean = meanPopWeight, sd = sdPopWeight)</code></p>
</div>
<div class="section level2">
<h2 id="add-the-two-together">add the two together<a class="anchor" aria-label="anchor" href="#add-the-two-together"></a>
</h2>
<p><code>pnorm(meanPopWeight - 1, mean = meanPopWeight, sd = sdPopWeight) +</code>
<code>1 - pnorm(meanPopWeight + 1, mean = meanPopWeight, sd = sdPopWeight)</code>
2. <code>set.seed(1)</code><br><code>n &lt;- 10000</code><br><code>meanSampleOf5 &lt;- vector("numeric", n)</code><br><code>for (i in 1:n) {</code><br><code>meanSampleOf5[i] &lt;- mean(sample(population$Bodyweight, 5))</code><br><code>}</code>meanSampleOf5<code></code>mean(population$Bodyweight)`</p>
</div>
<div class="section level2">
<h2 id="histogram-of-sample-means-1">histogram of sample means<a class="anchor" aria-label="anchor" href="#histogram-of-sample-means-1"></a>
</h2>
<p><code>hist(meanSampleOf5)</code></p>
</div>
<div class="section level2">
<h2 id="mean-population-weight-plus-one-gram-and-minus-one-gram-1">mean population weight plus one gram and minus one gram<a class="anchor" aria-label="anchor" href="#mean-population-weight-plus-one-gram-and-minus-one-gram-1"></a>
</h2>
<p><code>meanPopWeight &lt;- mean(population$Bodyweight)</code> # mean
<code>sdPopWeight &lt;- sd(population$Bodyweight)</code> # standard
deviation <code>abline(v = meanPopWeight, col = "blue", lwd = 2)</code>
<code>abline(v = meanPopWeight + 1, col = "red", lwd = 2)</code>
<code>abline(v = meanPopWeight - 1, col = "red", lwd = 2)</code></p>
</div>
<div class="section level2">
<h2 id="proportion-below-mean-population-weight-minus-1-gram-1">proportion below mean population weight minus 1 gram<a class="anchor" aria-label="anchor" href="#proportion-below-mean-population-weight-minus-1-gram-1"></a>
</h2>
<p><code>pnorm(meanPopWeight - 1, mean = meanPopWeight, sd = sdPopWeight)</code></p>
</div>
<div class="section level2">
<h2 id="proportion-greater-than-mean-population-weight-plus-1-gram-1">proportion greater than mean population weight plus 1 gram<a class="anchor" aria-label="anchor" href="#proportion-greater-than-mean-population-weight-plus-1-gram-1"></a>
</h2>
<p><code>1 - pnorm(meanPopWeight + 1, mean = meanPopWeight, sd = sdPopWeight)</code></p>
</div>
<div class="section level2">
<h2 id="add-the-two-together-1">add the two together<a class="anchor" aria-label="anchor" href="#add-the-two-together-1"></a>
</h2>
<p><code>pnorm(meanPopWeight - 1, mean = meanPopWeight, sd = sdPopWeight) +</code>
<code>1 - pnorm(meanPopWeight + 1, mean = meanPopWeight, sd = sdPopWeight)</code>
3. <code>set.seed(1)</code><br><code>n &lt;- 1000</code><br><code>meanSampleOf50 &lt;- vector("numeric", n)</code><br><code>for (i in 1:n) {</code><br><code>meanSampleOf50[i] &lt;- mean(sample(population$Bodyweight, 50))</code><br><code>}</code>meanSampleOf50<code></code>mean(population$Bodyweight)`</p>
</div>
<div class="section level2">
<h2 id="histogram-of-sample-means-2">histogram of sample means<a class="anchor" aria-label="anchor" href="#histogram-of-sample-means-2"></a>
</h2>
<p><code>hist(meanSampleOf50)</code></p>
</div>
<div class="section level2">
<h2 id="mean-population-weight-plus-one-gram-and-minus-one-gram-2">mean population weight plus one gram and minus one gram<a class="anchor" aria-label="anchor" href="#mean-population-weight-plus-one-gram-and-minus-one-gram-2"></a>
</h2>
<p><code>abline(v = meanPopWeight, col = "blue", lwd = 2)</code>
<code>abline(v = meanPopWeight + 1, col = "red", lwd = 2)</code>
<code>abline(v = meanPopWeight - 1, col = "red", lwd = 2)</code></p>
</div>
<div class="section level2">
<h2 id="proportion-below-mean-population-weight-minus-1-gram-2">proportion below mean population weight minus 1 gram<a class="anchor" aria-label="anchor" href="#proportion-below-mean-population-weight-minus-1-gram-2"></a>
</h2>
<p><code>pnorm(meanPopWeight - 1, mean = meanPopWeight, sd = sdPopWeight)</code></p>
</div>
<div class="section level2">
<h2 id="proportion-greater-than-mean-population-weight-plus-1-gram-2">proportion greater than mean population weight plus 1 gram<a class="anchor" aria-label="anchor" href="#proportion-greater-than-mean-population-weight-plus-1-gram-2"></a>
</h2>
<p><code>1 - pnorm(meanPopWeight + 1, mean = meanPopWeight, sd = sdPopWeight)</code></p>
</div>
<div class="section level2">
<h2 id="add-the-two-together-2">add the two together<a class="anchor" aria-label="anchor" href="#add-the-two-together-2"></a>
</h2>
<p><code>pnorm(meanPopWeight - 1, mean = meanPopWeight, sd = sdPopWeight) +</code>
<code>1 - pnorm(meanPopWeight + 1, mean = meanPopWeight, sd = sdPopWeight)</code></p>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Use a histogram to “look” at the distribution of averages we get with
a sample size of 5 and a sample size of 50. How would you say they
differ?<br>
A) They are actually the same.<br>
B) They both look roughly normal, but with a sample size of 50 the
spread is smaller.<br>
C) They both look roughly normal, but with a sample size of 50 the
spread is larger.<br>
D) The second distribution does not look normal at all.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Solution to Exercise 3
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>For the last set of averages, the ones obtained from a sample size of
50, what percent are between 23 and 25? Now ask the same question of a
normal distribution with average 23.9 and standard deviation 0.43. The
answer to the previous two were very similar. This is because we can
approximate the distribution of the sample average with a normal
distribution. We will learn more about the reason for this next.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Solution to Exercise 4
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Inference uses sampling to investigate population parameters such as
mean and standard deviation.</li>
<li>A p-value describes the probability of obtaining a specific
value.</li>
<li>A distribution summarizes a set of numbers.</li>
</ul>
</div>
</div>
</div>
</div>
</div></section><section id="aio-inference-pse"><p>Content from <a href="inference-pse.html">Populations, Samples and Estimates</a></p>
<hr>
<p>Last updated on 2025-08-17 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-pse.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is a parameter from a population?</li>
<li>What are sample estimates?</li>
<li>How can we use sample estimates to make inferences about population
parameters?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the difference between parameters and statistics.</li>
<li>Calculate population means and sample means.</li>
<li>Calculate the difference between the means of two subgroups.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="populations-samples-and-estimates">Populations, Samples and Estimates<a class="anchor" aria-label="anchor" href="#populations-samples-and-estimates"></a>
</h2>
<hr class="half-width">
<p>We can never know the true mean or variance of an entire population.
Why not? Because we can’t feasibly measure every member of a population.
We can never know the true mean blood pressure of all mice, for example,
even if all are from one strain, because we can’t afford to buy them all
or even find them all. We can never know the true mean blood pressure of
all people on a Western diet, for example, because we can’t possibly
measure the entire population that’s on a Western diet. If we could
measure all people on a Western diet, we really are interested in the
difference in means between people on a Western vs. non high fat high
sugar diet because we want to know what effect the diet has on people.
If there is no difference in means, we can say that there is no effect
of diet. If there is a difference in means, we can say that the diet has
an effect. The question we are asking can be expressed as:</p>
<p><img src="fig/02-muy-mux-eq-zero.png" alt="Does mean of Y less mean of X equal zero?" class="figure"> We can even compare
more than two means. The three normal curves below help to visualize a
question comparing the means of each curve to one of the others.</p>
<figure><img src="fig/normal-curves.png" alt="Normal curves with different means and standard deviations" class="figure mx-auto d-block"></figure><p>We also want to know the variance from the mean, so that we have a
sense of the spread of measurement around the mean.</p>
<p>In reality we use sample estimates of population parameters. The true
population parameters that we are interested in are mean and standard
deviation. Here we learn how taking a sample permits us to answer our
questions about differences between groups. This is the essence of
statistical inference.</p>
<p>Now that we have introduced the idea of a random variable, a null
distribution, and a p-value, we are ready to describe the mathematical
theory that permits us to compute p-values in practice. We will also
learn about confidence intervals and power calculations.</p>
<div class="section level4">
<h4 id="population-parameters">Population parameters<a class="anchor" aria-label="anchor" href="#population-parameters"></a>
</h4>
<p>A first step in statistical inference is to understand what
population you are interested in. In the mouse weight example, we have
two populations: female mice on control diets and female mice on high
fat diets, with weight being the outcome of interest. We consider this
population to be fixed, and the randomness comes from the sampling. One
reason we have been using this dataset as an example is because we
happen to have the weights of all the mice of this type. We download <a href="https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/mice_pheno.csv" class="external-link">this</a>
file to our working directory and read in to R:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/mice_pheno.csv"</span><span class="op">)</span></span></code></pre>
</div>
<p>We can then access the population values and determine, for example,
how many we have. Here we compute the size of the control
population:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">controlPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 225</code></pre>
</div>
<p>We usually denote these values as <i>x<sub>
1</sub>,…,x<sub>m</sub></i>. In this case, <em>m</em> is the number
computed above. We can do the same for the high fat diet population:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hfPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 200</code></pre>
</div>
<p>and denote with <i>y<sub> 1</sub>,…,y<sub>n</sub></i>.</p>
<p>We can then define summaries of interest for these populations, such
as the mean and variance.</p>
<p>The mean:</p>
<figure><img src="fig/02-mux-and-muy.png" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># X is the control population</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span> <span class="co"># sum of the xsubi's</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 5376.01</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span> <span class="co"># this equals m</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 225</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span> <span class="co"># this equals mu sub x</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23.89338</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Y is the high fat diet population</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span> <span class="co"># sum of the ysubi's</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 5253.779</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span> <span class="co"># this equals mu sub y</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 26.2689</code></pre>
</div>
<p>The variance:</p>
<figure><img src="fig/02-var-x-and-y.png" class="figure mx-auto d-block"></figure><p>with the standard deviation being the square root of the variance. We
refer to such quantities that can be obtained from the population as
<em>population parameters</em>. The question we started out asking can
now be written mathematically: <strong>is <i>μ<sub>Y</sub> -
μ<sub>X</sub> = 0</i></strong> ?</p>
<p>Although in our illustration we have all the values and can check if
this is true, in practice we do not. For example, in practice it would
be prohibitively expensive to buy all the mice in a population. Here we
learn how taking a <em>sample</em> permits us to answer our questions.
<strong><i>This is the essence of statistical
inference.</i></strong></p>
</div>
<div class="section level4">
<h4 id="sample-estimates">Sample estimates<a class="anchor" aria-label="anchor" href="#sample-estimates"></a>
</h4>
<p>In the previous chapter, we obtained samples of 12 mice from each
population. We represent data from samples with capital letters to
indicate that they are random. This is common practice in statistics,
although it is not always followed. So the samples are <i>X<sub>
1</sub>,…,X<sub>M</sub></i> and <i>Y<sub> 1</sub>,…,Y<sub>N</sub></i>
and, in this case, <i>Ν=Μ=12</i>. In contrast and as we saw above, when
we list out the values of the population, which are set and not random,
we use lower-case letters.</p>
<p>Since we want to know if <i>μ<sub>Y</sub> - μ<sub>X</sub> = 0</i>, we
consider the sample version: Ȳ-X̄ with:</p>
<p><img src="fig/XbarEquals.png" class="figure"><img src="fig/andYbarEquals.png" class="figure"></p>
<p>Note that this difference of averages is also a random variable.
Previously, we learned about the behavior of random variables with an
exercise that involved repeatedly sampling from the original
distribution. Of course, this is not an exercise that we can execute in
practice. In this particular case it would involve buying 24 mice over
and over again. Here we described the mathematical theory that
mathematically relates X̄ to <i>μ<sub>X</sub></i> and Ȳ to
<i>μ<sub>Y</sub></i>, that will in turn help us understand the
relationship between Ȳ-X̄ and <i>μ<sub>Y</sub> - μ<sub>X</sub></i>.
Specifically, we will describe how the Central Limit Theorem permits us
to use an approximation to answer this question, as well as motivate the
widely used t-distribution.</p>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>We will use the mouse phenotype data. Remove the observations that
contain missing values:</p>
<p><code>pheno &lt;- na.omit(pheno)</code></p>
<p>Use <code>dplyr</code> to create a vector <code>x</code> with the
body weight of all males on the control (chow) diet.<br>
What is this population’s average body weight?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Omit observations with missing data</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu">na.omit</span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create subset of males on chow diet</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">pheno</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">Diet</span> <span class="op">==</span> <span class="st">"chow"</span> <span class="op">&amp;</span> <span class="va">Sex</span> <span class="op">==</span> <span class="st">"M"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="va">unlist</span></span>
<span></span>
<span><span class="co"># Calculate mean weight</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 30.96381</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<em> (continued)</em>
</h3>
<div class="callout-content">
<p>Set the seed at 1. Take a random sample <code>X</code> of size 25
from <code>x</code>. What is the sample average?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">x_sample</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">x</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">x_sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 30.5196</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<em> (continued)</em>
</h3>
<div class="callout-content">
<p>Use <code>dplyr</code> to create a vector <code>y</code> with the
body weight of all males on the high fat (hf) diet.<br>
What is this population’s average?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create subset of males on high fat diet</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">pheno</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">Diet</span> <span class="op">==</span> <span class="st">"hf"</span> <span class="op">&amp;</span> <span class="va">Sex</span> <span class="op">==</span> <span class="st">"M"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="va">unlist</span></span>
<span></span>
<span><span class="co"># Calculate the mean weight</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 34.84793</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<em> (continued)</em>
</h3>
<div class="callout-content">
<p>Set the seed at 1. Take a random sample <code>Y</code> of size 25
from <code>y</code>. What is the sample average?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y_sample</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">y</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y_sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 35.8036</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<em> (continued)</em>
</h3>
<div class="callout-content">
<p>What is the difference in absolute value between <i>x̄-ȳ</i> and
X̄-Ȳ?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># mu_x - mu_y</span></span>
<span><span class="va">pop_mean_diff</span> <span class="op">&lt;-</span> <span class="fu">abs</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># x-bar - y-bar</span></span>
<span><span class="va">sample_mean_diff</span> <span class="op">&lt;-</span> <span class="fu">abs</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">x_sample</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y_sample</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># difference in absolute value between (mu_x - mu_y) - (x-bar - y-bar)</span></span>
<span><span class="va">pop_sample_mean_diff</span> <span class="op">&lt;-</span> <span class="fu">abs</span><span class="op">(</span><span class="va">pop_mean_diff</span> <span class="op">-</span> <span class="va">sample_mean_diff</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pop_sample_mean_diff</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.399884</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise<em> (continued)</em>
</h3>
<div class="callout-content">
<p>Repeat the above for females. Make sure to set the seed to 1 before
each sample call. What is the difference in absolute value between
<i>x̄-ȳ</i> and X̄-Ȳ?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x_f</span> <span class="op">&lt;-</span> <span class="va">pheno</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">Diet</span> <span class="op">==</span> <span class="st">"chow"</span> <span class="op">&amp;</span> <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="va">unlist</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">x_f</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23.89338</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">x_f_sample</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">x_f</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">x_f_sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 24.2528</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y_f</span> <span class="op">&lt;-</span> <span class="va">pheno</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">Diet</span> <span class="op">==</span> <span class="st">"hf"</span> <span class="op">&amp;</span> <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="va">unlist</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y_f</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 26.2689</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y_f_sample</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">y_f</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y_f_sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 28.3828</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pop_mean_diff_f</span> <span class="op">&lt;-</span> <span class="fu">abs</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">x_f</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y_f</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sample_mean_diff_f</span> <span class="op">&lt;-</span> <span class="fu">abs</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">x_f_sample</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y_f_sample</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pop_sample_mean_diff_f</span> <span class="op">&lt;-</span> <span class="fu">abs</span><span class="op">(</span><span class="va">pop_mean_diff_f</span> <span class="op">-</span> <span class="va">sample_mean_diff_f</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pop_sample_mean_diff_f</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.754483</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Parameters are measures of populations.</li>
<li>Statistics are measures of samples drawn from populations.</li>
<li>Statistical inference is the process of using sample statistics to
answer questions about population parameters.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-inference-clt"><p>Content from <a href="inference-clt.html">Central Limit Theorem and the t-distribution</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-clt.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is a parameter from a population?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<hr>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="central-limit-theorem-and-t-distribution">Central Limit Theorem and t-distribution<a class="anchor" aria-label="anchor" href="#central-limit-theorem-and-t-distribution"></a>
</h2>
<hr class="half-width">
<p>Below we will discuss the Central Limit Theorem (CLT) and the
t-distribution, both of which help us make important calculations
related to probabilities. Both are frequently used in science to test
statistical hypotheses. To use these, we have to make different
assumptions from those for the CLT and the t-distribution. However, if
the assumptions are true, then we are able to calculate the exact
probabilities of events through the use of mathematical formula.</p>
<div class="section level4">
<h4 id="central-limit-theorem">Central Limit Theorem<a class="anchor" aria-label="anchor" href="#central-limit-theorem"></a>
</h4>
<p>The CLT is one of the most frequently used mathematical results in
science. It tells us that when the sample size is large, the average Ȳ
of a random sample follows a normal distribution centered at the
population average μ<sub>Y</sub> and with standard deviation equal to
the population standard deviation σ<sub>Y</sub>, divided by the square
root of the sample size <em>N</em>. We refer to the standard deviation
of the distribution of a random variable as the random variable’s
<em>standard error</em>.</p>
<p>This implies that if we take many samples of size <em>N</em>, then
the quantity:</p>
<figure><img src="fig/02-random-sample.png" class="figure mx-auto d-block"></figure><p>is approximated with a normal distribution centered at 0 and with
standard deviation 1.</p>
<p>We are interested in the difference between two sample averages.
Again, applying certain mathematical principles, it can be implied that
the below ratio:</p>
<figure><img src="fig/02-diff-sample-avgs.png" class="figure mx-auto d-block"></figure><p>is approximated by a normal distribution centered at 0 and standard
deviation 1. Calculating p-values for the standard normal distribution
is simple because we know the proportion of the distribution under any
value. For example, only 5% of the values in the standard normal
distribution are larger than 2 (in absolute value):</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pnorm</span><span class="op">(</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu">pnorm</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.04550026</code></pre>
</div>
<p>We don’t need to buy more mice, 12 and 12 suffice.</p>
<p>However, we can’t claim victory just yet because we don’t know the
population standard deviations: σ<sub>X</sub> and σ<sub>Y</sub>. These
are unknown population parameters, but we can get around this by using
the sample standard deviations, call them <em>s<sub>X</sub></em> and
<em>s<sub>Y</sub></em>. These are defined as:</p>
<figure><img src="fig/02-sample-stddev.png" class="figure mx-auto d-block"></figure><p>Note that we are dividing by <em>M-1</em> and <em>N-1</em>, instead
of by <em>M</em> and <em>N</em>. There is a theoretical reason for doing
this which we do not explain here.</p>
<p>So we can redefine our ratio as</p>
<figure><img src="fig/02-sqrtN.png" class="figure mx-auto d-block"></figure><p>if <em>M = N</em> or in general,</p>
<figure><img src="fig/02-ybar-xbar.png" class="figure mx-auto d-block"></figure><p>The CLT tells us that when <em>M</em> and <em>N</em> are large, this
random variable is normally distributed with mean 0 and SD 1. Thus we
can compute p-values using the function <code>pnorm</code>.</p>
</div>
<div class="section level4">
<h4 id="the-t-distribution">The t-distribution<a class="anchor" aria-label="anchor" href="#the-t-distribution"></a>
</h4>
<p>The CLT relies on large samples, what we refer to as <em>asymptotic
results</em>. When the CLT does not apply, there is another option that
does not rely on asymptotic results. When the original population from
which a random variable, say <em>Y</em>, is sampled is normally
distributed with mean 0, then we can calculate the distribution of:</p>
<figure><img src="fig/02-tdist.gif" class="figure mx-auto d-block"></figure><p>This is the ratio of two random variables so it is not necessarily
normal. The fact that the denominator can be small by chance increases
the probability of observing large values. <a href="https://en.wikipedia.org/wiki/William_Sealy_Gosset" class="external-link">William Sealy
Gosset</a>, an employee of the Guinness brewing company, deciphered the
distribution of this random variable and published a paper under the
pseudonym “Student”. The distribution is therefore called Student’s
t-distribution. Later we will learn more about how this result is
used.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>If a list of numbers has a distribution that is well approximated by
the normal distribution, what proportion of these numbers are within one
standard deviation away from the list’s average?</li>
<li>What proportion of these numbers are within two standard deviations
away from the list’s average?</li>
<li>What proportion of these numbers are within three standard
deviations away from the list’s average?</li>
<li>Define y to be the weights of males on the control diet. What
proportion of the mice are within one standard deviation away from the
average weight (remember to use <code>popsd</code> for the population
sd)?</li>
<li>What proportion of these numbers are within two standard deviations
away from the list’s average?</li>
<li>What proportion of these numbers are within three standard
deviations away from the list’s average?</li>
<li>Note that the numbers for the normal distribution and our weights
are relatively close. Also, notice that we are indirectly comparing
quantiles of the normal distribution to quantiles of the mouse weight
distribution. We can actually compare all quantiles using a qqplot.
Which of the following best describes the qq-plot comparing mouse
weights to the normal distribution?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>The points on the qq-plot fall exactly on the identity line.<br>
</li>
<li>The average of the mouse weights is not 0 and thus it can’t follow a
normal distribution.<br>
</li>
<li>The mouse weights are well approximated by the normal distribution,
although the larger values (right tail) are larger than predicted by the
normal. This is consistent with the differences seen between question 3
and 6.<br>
</li>
<li>These are not random variables and thus they can’t follow a normal
distribution.</li>
</ol>
<ol start="8" style="list-style-type: decimal">
<li>Create the above qq-plot for the four populations: male/females on
each of the two diets. What is the most likely explanation for the mouse
weights being well approximated?<br>
What is the best explanation for all these being well approximated by
the normal distribution?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>The CLT tells us that sample averages are approximately
normal.<br>
</li>
<li>This just happens to be how nature behaves, perhaps the result of
many biological factors averaging out.<br>
</li>
<li>Everything measured in nature follows a normal distribution.<br>
</li>
<li>Measurement error is normally distributed.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Solutions to Exercise 1
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>
<code>pnorm(1)-pnorm(-1)</code> = 0.6826895 (68%)</li>
<li>
<code>pnorm(2)-pnorm(-2)</code> = 0.9544997 (95%)</li>
<li>
<code>pnorm(3)-pnorm(-3)</code> = 0.9973002 (99%)</li>
<li>
<code>install.packages("rafalib")</code><br><code>library(rafalib)</code><br><code>pheno &lt;- read.csv(file = "../data/mice_pheno.csv")</code><br><code>y &lt;- pheno %&gt;%</code><br><code>filter(Sex == "M" &amp; Diet == "chow") %&gt;%</code><br><code>select(Bodyweight) %&gt;%</code><br><code>unlist()</code><br><code>sigma &lt;- popsd(y, na.rm = TRUE)</code> # 4.420545<br><code>mu &lt;- mean(y, na.rm = TRUE)</code> # 30.96381<br><code>y %&gt;%</code><br><code>filter(Bodyweight &gt; mu - sigma &amp; Bodyweight &lt; mu + sigma) %&gt;%</code><br><code>count()/nrow(y)</code> # 0.6919643 (~68%)<br><code>y %&gt;%</code><br><code>filter(Bodyweight &gt; mu - 2*sigma &amp; Bodyweight &lt; mu + 2*sigma) %&gt;%</code><br><code>count()/nrow(y)</code> # 0.9419643 (~95%)<br><code>filter(Bodyweight &gt; mu - 3*sigma &amp; Bodyweight &lt; mu + 3*sigma) %&gt;%</code><br><code>count()/nrow(y)</code> # 0.9866071 (~99%)</li>
</ol>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Here we are going to use the function <code>replicate</code> to learn
about the distribution of random variables. All the above exercises
relate to the normal distribution as an approximation of the
distribution of a fixed list of numbers or a population. We have not yet
discussed probability in these exercises. If the distribution of a list
of numbers is approximately normal, then if we pick a number at random
from this distribution, it will follow a normal distribution. However,
it is important to remember that stating that some quantity has a
distribution does not necessarily imply this quantity is random. Also,
keep in mind that this is not related to the central limit theorem. The
central limit applies to averages of random variables. Let’s explore
this concept.<br>
We will now take a sample of size 25 from the population of males on the
chow diet. The average of this sample is our random variable. We will
use the replicate to observe 10,000 realizations of this random
variable.</p>
<ol style="list-style-type: decimal">
<li>Set the seed at 1, generate these 10,000 averages.<br><code>set.seed(1)</code><br><code>y &lt;- filter(pheno, Sex=="M" &amp; Diet=="chow") %&gt;%</code><br><code>select(Bodyweight) %&gt;%</code><br><code>unlist</code><br><code>avgs &lt;- replicate(10000, mean(sample(y,25)))</code>
</li>
<li>Make a histogram and qq-plot these 10,000 numbers against the normal
distribution. <code>mypar(1,2)</code><br><code>hist(avgs)</code><br><code>qqnorm(avgs)</code><br><code>qqline(avgs)</code><br>
We can see that, as predicted by the CLT, the distribution of the random
variable is very well approximated by the normal distribution.</li>
<li>What is the average of the distribution of the sample average?</li>
<li>What is the standard deviation of the distribution of sample
averages?<br>
According to the CLT, the answer to the exercise above should be the
same as <code>mean(y)</code>. You should be able to confirm that these
two numbers are very close.</li>
<li>Which of the following does the CLT tell us should be close to your
answer to this exercise?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>
<code>popsd(y)</code><br>
</li>
<li>
<code>popsd(avgs)/sqrt(25)</code><br>
</li>
<li>
<code>sqrt(25) / popsd(y)</code><br>
</li>
<li><code>popsd(y)/sqrt(25)</code></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Solution to Exercise 2
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<p>Here we will use the mice phenotype data as an example. We start by
creating two vectors, one for the control population and one for the
high-fat diet population:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pheno &lt;- read.csv("mice_pheno.csv") #We downloaded this file in a previous section</span></span>
<span><span class="va">controlPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span><span class="va">hfPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span></code></pre>
</div>
<p>It is important to keep in mind that what we are assuming to be
normal here is the distribution of
<i>y<sub>1</sub>,y<sub>2</sub>,…,y<sub>n</sub></i>, not the random
variable Ȳ. Although we can’t do this in practice, in this illustrative
example, we get to see this distribution for both controls and high fat
diet mice:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-clt-rendered-population_histograms-1.png" alt="Histograms of all weights for both populations." class="figure mx-auto d-block"><figcaption>
Histograms of all weights for both populations.
</figcaption></figure><p>We can use <em>qq-plots</em> to confirm that the distributions are
relatively close to being normally distributed. We will explore these
plots in more depth in a later section, but the important thing to know
is that it compares data (on the y-axis) against a theoretical
distribution (on the x-axis). If the points fall on the identity line,
then the data is close to the theoretical distribution.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-clt-rendered-population_qqplots-1.png" alt="Quantile-quantile plots of all weights for both populations." class="figure mx-auto d-block"><figcaption>
Quantile-quantile plots of all weights for both populations.
</figcaption></figure><p>The larger the sample, the more forgiving the result is to the
weakness of this approximation. In the next section, we will see that
for this particular dataset the t-distribution works well even for
sample sizes as small as 3.</p>
<p>All the above exercises relate to the normal distribution as an
approximation of the distribution of a fixed list of numbers or a
population. We have not yet discussed probability in these exercises. If
the distribution of a list of numbers is approximately normal, then if
we pick a number at random from this distribution, it will follow a
normal distribution. However, it is important to remember that stating
that some quantity has a distribution does not necessarily imply this
quantity is random. Also, keep in mind that this is not related to the
central limit theorem. The central limit applies to averages of random
variables. Let’s explore this concept.</p>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>We will now take a sample of size 25 from the population of males on
the chow diet. The average of this sample is our random variable. We
will use <code>replicate</code> to observe 10,000 realizations of this
random variable.</p>
<ol style="list-style-type: decimal">
<li>Set the seed at 1.</li>
<li>Generate these 10,000 averages.</li>
<li>Make a histogram and qq-plot of these 10,000 numbers against the
normal distribution.<br><code>y &lt;- filter(dat, Sex=="M" &amp; Diet=="chow") %&gt;%</code><br><code>select(Bodyweight) %&gt;%</code><br><code>unlist</code><br><code>avgs &lt;- replicate(10000, mean( sample(y,25)))</code><br><code>mypar(1,2)</code><br><code>hist(avgs)</code><br><code>qqnorm(avgs)</code><br><code>qqline(avgs)</code><br>
We can see that, as predicted by the CLT, the distribution of the random
variable is very well approximated by the normal distribution.</li>
<li>What is the average of the distribution of the sample average?</li>
<li>What is the standard deviation of the distribution of sample
averages?</li>
<li>According to the CLT, the answer to exercise 9 should be the same as
<code>mean(y)</code>. You should be able to confirm that these two
numbers are very close. Which of the following does the CLT tell us
should be close to your answer to exercise 5?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>
<code>popsd(y)</code><br>
</li>
<li>
<code>popsd(avgs) / sqrt(25)</code><br>
</li>
<li>
<code>sqrt(25) / popsd(y)</code><br>
</li>
<li><code>popsd(y) / sqrt(25)</code></li>
</ol>
<ol start="7" style="list-style-type: decimal">
<li>In practice we do not know <code>(popsd(y))</code> which is why we
can’t use the CLT directly. This is because we see a sample and not the
entire distribution. We also can’t use <code>popsd(avgs)</code> because
to construct averages, we have to take 10,000 samples and this is never
practical. We usually just get one sample. Instead we have to estimate
<code>popsd(y)</code>. As described, what we use is the sample standard
deviation.<br>
Set the seed at 1. Using the<code>replicate</code> function, create
10,000 samples of 25 and now, instead of the sample average, keep the
standard deviation. Look at the distribution of the sample standard
deviations. It is a random variable. The real population SD is about
4.5. What proportion of the sample SDs are below 3.5?</li>
<li>What the answer to question 7 reveals is that the denominator of the
t-test is a random variable. By decreasing the sample size, you can see
how this variability can increase. It therefore adds variability. The
smaller the sample size, the more variability is added. The normal
distribution stops providing a useful approximation. When the
distribution of the population values is approximately normal, as it is
for the weights, the t-distribution provides a better approximation. We
will see this later on. Here we will look at the difference between the
t-distribution and normal. Use the function <code>qt</code> and
<code>qnorm</code> to get the quantiles of
<code>x=seq(0.0001, 0.9999, len=300)</code>. Do this for degrees of
freedom 3, 10, 30, and 100. Which of the following is true?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>The t-distribution and normal distribution are always the
same.<br>
</li>
<li>The t-distribution has a higher average than the normal
distribution.<br>
</li>
<li>The t-distribution has larger tails up until 30 degrees of freedom,
at which point itis practically the same as the normal
distribution.<br>
</li>
<li>The variance of the t-distribution grows as the degrees of freedom
grow.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Solution to Exercise 3
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>We can calculate the exact probabilities of events through using
mathematical formulas.</li>
<li>The Central Limit Theorem and t-distribution have different
assumptions.</li>
<li>Both can be used to calculate probabilities.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-inference-clt-practice"><p>Content from <a href="inference-clt-practice.html">Central Limit Theorem in practice</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-clt-practice.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How is the CLT used in practice?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<hr>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="central-limit-theorem-in-practice">Central Limit Theorem in Practice<a class="anchor" aria-label="anchor" href="#central-limit-theorem-in-practice"></a>
</h2>
<hr class="half-width">
<p>Let’s use our data to see how well the central limit theorem
approximates sample averages from our data. We will leverage our entire
population dataset to compare the results we obtain by actually sampling
from the distribution to what the CLT predicts.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># pheno &lt;- read.csv("mice_pheno.csv") #file was previously downloaded</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Sex Diet Bodyweight
1   F   hf      31.94
2   F   hf      32.48
3   F   hf      22.82
4   F   hf      19.92
5   F   hf      32.22
6   F   hf      27.50</code></pre>
</div>
<p>Start by selecting only female mice since males and females have
different weights. We will select three mice from each population.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="va">controlPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span><span class="va">hfPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span></code></pre>
</div>
<p>We can compute the population parameters of interest using the mean
function.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mu_hf</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span></span>
<span><span class="va">mu_control</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">mu_hf</span> <span class="op">-</span> <span class="va">mu_control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2.375517</code></pre>
</div>
<p>We can compute the population standard deviations of, say, a vector
<i>x</i> as well. However, we do not use the R function <code>sd</code>
because this function actually does not compute the population standard
deviation <i>σ<sub>x</sub></i>. Instead, <code>sd</code> assumes the
main argument is a random sample, say <i>X</i>, and provides an estimate
of <i>σ<sub>x</sub></i>, defined by <i>s<sub>X</sub></i> above. As shown
in the equations above the actual final answer differs because one
divides by the sample size and the other by the sample size minus one.
We can see that with R code:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">controlPopulation</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="va">populationvar</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="op">(</span><span class="va">x</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">identical</span><span class="op">(</span><span class="fu">var</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="va">populationvar</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] FALSE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">identical</span><span class="op">(</span><span class="fu">var</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="va">N</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="va">N</span>, <span class="va">populationvar</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<p>So to be mathematically correct, we do not use <code>sd</code> or
<code>var</code>. Instead, we use the <code>popvar</code> and
<code>popsd</code> function in <code>rafalib</code>:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="va">sd_hf</span> <span class="op">&lt;-</span> <span class="fu">popsd</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span></span>
<span><span class="va">sd_control</span> <span class="op">&lt;-</span> <span class="fu">popsd</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span></code></pre>
</div>
<p>Remember that in practice we do not get to compute these population
parameters. These are values we never see. In general, we want to
estimate them from samples.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">hf</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">hfPopulation</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>, <span class="fl">12</span><span class="op">)</span></span></code></pre>
</div>
<p>As we described, the CLT tells us that for large <i>N</i>, each of
these is approximately normal with average population mean and standard
error population variance divided by <i>N</i>. We mentioned that a rule
of thumb is that <i>N</i> should be 30 or more. However, that is just a
rule of thumb since the preciseness of the approximation depends on the
population distribution. Here we can actually check the approximation
and we do that for various values of <i>N</i>.</p>
<p>Now we use <code>sapply</code> and <code>replicate</code> instead of
<code>for</code> loops, which makes for cleaner code (we do not have to
pre-allocate a vector, R takes care of this for us):</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ns</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">12</span>,<span class="fl">25</span>,<span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">10000</span> <span class="co">#number of simulations</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span>  <span class="fu">sapply</span><span class="op">(</span><span class="va">Ns</span>,<span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>,<span class="fu">mean</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">hfPopulation</span>,<span class="va">n</span><span class="op">)</span><span class="op">)</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>,<span class="va">n</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can use qq-plots to see how well CLT approximations works for
these. If in fact the normal distribution is a good approximation, the
points should fall on a straight line when compared to normal quantiles.
The more it deviates, the worse the approximation. In the title, we also
show the average and SD of the observed distribution, which demonstrates
how the SD decreases with <i>√N</i> as predicted.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span>along<span class="op">=</span><span class="va">Ns</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">titleavg</span> <span class="op">&lt;-</span> <span class="fu">signif</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span>
<span>  <span class="va">titlesd</span> <span class="op">&lt;-</span> <span class="fu">signif</span><span class="op">(</span><span class="fu">popsd</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span>
<span>  <span class="va">title</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">"N="</span>, <span class="va">Ns</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,<span class="st">" Avg="</span>, <span class="va">titleavg</span>,<span class="st">" SD="</span>, <span class="va">titlesd</span><span class="op">)</span></span>
<span>  <span class="fu">qqnorm</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, main<span class="op">=</span><span class="va">title</span><span class="op">)</span></span>
<span>  <span class="fu">qqline</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, col<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-clt-practice-rendered-effect_size_qqplot-1.png" alt="Quantile versus quantile plot of simulated differences versus theoretical normal distribution for four different sample sizes." class="figure mx-auto d-block"><figcaption>
Quantile versus quantile plot of simulated differences versus
theoretical normal distribution for four different sample sizes.
</figcaption></figure><p>Here we see a pretty good fit even for 3. Why is this? Because the
population itself is relatively close to normally distributed, the
averages are close to normal as well (the sum of normals is also a
normal). In practice, we actually calculate a ratio: we divide by the
estimated standard deviation. Here is where the sample size starts to
matter more.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ns</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>, <span class="fl">12</span>, <span class="fl">25</span>, <span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">10000</span> <span class="co">#number of simulations</span></span>
<span><span class="co">##function to compute a t-stat</span></span>
<span><span class="va">computetstat</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">hfPopulation</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="fu">var</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">/</span><span class="va">n</span> <span class="op">+</span> <span class="fu">var</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">/</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span>  <span class="fu">sapply</span><span class="op">(</span><span class="va">Ns</span>, <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="fu">computetstat</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span>along<span class="op">=</span><span class="va">Ns</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">qqnorm</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, main<span class="op">=</span><span class="va">Ns</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu">qqline</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, col<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-clt-practice-rendered-t_test_qqplot-1.png" alt="Quantile versus quantile plot of simulated ratios versus theoretical normal distribution for four different sample sizes." class="figure mx-auto d-block"><figcaption>
Quantile versus quantile plot of simulated ratios versus theoretical
normal distribution for four different sample sizes.
</figcaption></figure><p>So we see that for <i>N = 3</i>, the CLT does not provide a usable
approximation. For <i>N = 12</i>, there is a slight deviation at the
higher values, although the approximation appears useful. For 25 and 50,
the approximation is spot on.</p>
<p>This simulation only proves that <i>N = 12</i> is large enough in
this case, not in general. As mentioned above, we will not be able to
perform this simulation in most situations. We only use the simulation
to illustrate the concepts behind the CLT and its limitations. In future
sections, we will describe the approaches we actually use in
practice.</p>
<p>These exercises use the female mouse weights data set we have
previously downloaded.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>The CLT is a result from probability theory. Much of probability
theory was originally inspired by gambling. This theory is still used in
practice by casinos. For example, they can estimate how many people need
to play slots for there to be a 99.9999% probability of earning enough
money to cover expenses. Let’s try a simple example related to gambling.
Suppose we are interested in the proportion of times we see a 6 when
rolling n=100 die. This is a random variable which we can simulate with
<code>x=sample(1:6, n, replace=TRUE)</code> and the proportion we are
interested in can be expressed as an average: <code>mean(x==6)</code>.
Because the die rolls are independent, the CLT applies. We want to roll
<i>n</i> dice 10,000 times and keep these proportions. This random
variable (proportion of 6s) has mean <i>p=1/6</i> and variance <i>p ×
(1-p)/n</i>. So according to CLT <i>z</i> =
<code>(mean(x==6) - p) / sqrt(p * (1-p)/n)</code> should be normal with
mean 0 and SD 1.<br>
Set the seed to 1, then use <code>replicate</code> to perform the
simulation, and report what proportion of times z was larger than 2 in
absolute value (CLT says it should be about 0.05).</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Solution to Exercise 1
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>For the last simulation you can make a <code>qqplot</code> to confirm
the normal approximation. Now, the CLT is an asympototic result, meaning
it is closer and closer to being a perfect approximation as the sample
size increases. In practice, however, we need to decide if it is
appropriate for actual sample sizes. Is 10 enough? 15? 30? In the
example used in exercise 1, the original data is binary (either 6 or
not). In this case, the success probability also affects the
appropriateness of the CLT. With very low probabilities, we need larger
sample sizes for the CLT to “kick in”. Run the simulation from exercise
1, but for different values of <i>p</i> and <i>n</i>. For which of the
following is the normal approximation best?<br>
A) <i>p</i>=0.5 and <i>n</i>=5<br>
B) <i>p</i>=0.5 and <i>n</i>=30<br>
C) <i>p</i>=0.01 and <i>n</i>=30<br>
D) <i>p</i>=0.01 and <i>n</i>=100</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Solution to Exercise 2
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>As we have already seen, the CLT also applies to averages of
quantitative data. A major difference with binary data, for which we
know the variance is <i>p(1 - p)</i>, is that with quantitative data we
need to estimate the population standard deviation. In several previous
exercises we have illustrated statistical concepts with the unrealistic
situation of having access to the entire population. In practice, we do
not have access to entire populations. Instead, we obtain one random
sample and need to reach conclusions analyzing that data.
<code>pheno</code> is an example of a typical simple dataset
representing just one sample. We have 12 measurements for each of two
populations:</p>
<p><code>X &lt;- filter(dat, Diet=="chow") %&gt;%</code><br><code>select(Bodyweight) %&gt;%</code><br><code>unlist</code><br><code>Y &lt;- filter(dat, Diet=="hf") %&gt;%</code><br><code>select(Bodyweight) %&gt;%</code><br><code>unlist</code><br>
We think of <i>X</i> as a random sample from the population of all mice
in the control diet and <i>Y</i> as a random sample from the population
of all mice in the high fat diet.</p>
<ol style="list-style-type: decimal">
<li>Define the parameter <i>μ<sub>x</sub></i> as the average of the
control population. We estimate this parameter with the sample average
X̄. What is the sample average?</li>
<li>We don’t know <i>μ<sub>X</sub></i>, but want to use X̄ to understand
<i>μ<sub>X</sub></i>. Which of the following uses CLT to understand how
well X̄ approximates <i>μ<sub>X</sub></i>?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>X̄ follows a normal distribution with mean 0 and standard deviation
1.<br>
</li>
<li>
<i>μ<sub>X</sub></i> follows a normal distribution with mean X̄ and
standard deviation <i>σ<sub>X</sub>/√12</i> where <i>σ<sub>X</sub></i>
is the population standard deviation.<br>
</li>
<li>X̄ follows a normal distribution with mean <i>μ<sub>X</sub></i> and
standard deviation <i>σ<sub>X</sub></i> where <i>σ<sub>X</sub></i> is
the population standard deviation.<br>
</li>
<li>X̄ follows a normal distribution with mean <i>μ<sub>X</sub></i> and
standard deviation <i>σ<sub>X</sub>/√12</i> where <i>σ<sub>X</sub></i>
is the population standard deviation.</li>
</ol>
<ol start="3" style="list-style-type: decimal">
<li>The result above tells us the distribution of the following random
variable:<br><img src="fig/Z-equation.png" class="figure"> What does the CLT tell us is the mean
of <i>Z</i> (you don’t need code)?</li>
<li>The result of 2 and 3 tell us that we know the distribution of the
difference between our estimate and what we want to estimate, but don’t
know. However, the equation involves the population standard deviation
<i>σ<sub>X</sub></i>, which we don’t know. Given what we discussed, what
is your estimate of <i>σ<sub>X</sub></i>?</li>
<li>Use the CLT to approximate the probability that our estimate X̄ is
off by more than 5.21 ounces from <i>μ<sub>X</sub></i>.</li>
<li>Now we introduce the concept of a null hypothesis. We don’t know
<i>μ<sub>x</sub></i> nor <i>μ<sub>y</sub></i>. We want to quantify what
the data say about the possibility that the diet has no effect:
<i>μ<sub>x</sub> = μ<sub>y</sub></i>. If we use CLT, then we approximate
the distribution of X̄ as normal with mean <i>μ<sub>X</sub></i> and
standard deviation <i>σ<sub>X</sub></i> and the distribution of Ȳ as
normal with mean <i>μ<sub>Y</sub></i> and standard deviation
<i>σ<sub>Y</sub></i>. This implies that the difference Ȳ - X̄ has mean 0.
We described that the standard deviation of this statistic (the standard
error) is <img src="fig/SE-Xbar-Ybar.png" class="figure"> and that we estimate the
population standard deviations <i>σ<sub>X</sub></i> and
<i>σ<sub>Y</sub></i> with the sample estimates. What is the estimate of
<img src="fig/SE-Xbar-Ybar.png" class="figure">
</li>
<li>So now we can compute Ȳ - X̄ as well as an estimate of this standard
error and construct a t-statistic. What is this t-statistic?</li>
<li>If we apply the CLT, what is the distribution of this
t-statistic?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Normal with mean 0 and standard deviation 1.<br>
</li>
<li>t-distributed with 22 degrees of freedom.<br>
</li>
<li>Normal with mean 0 and standard deviation <img src="fig/SE.png" class="figure"><br>
</li>
<li>t-distributed with 12 degrees of freedom.</li>
</ol>
<ol start="9" style="list-style-type: decimal">
<li>Now we are ready to compute a p-value using the CLT. What is the
probability of observing a quantity as large as what we computed in 8,
when the null distribution is true?</li>
<li>CLT provides an approximation for cases in which the sample size is
large. In practice, we can’t check the assumption because we only get to
see 1 outcome (which you computed above). As a result, if this
approximation is off, so is our p-value. As described earlier, there is
another approach that does not require a large sample size, but rather
that the distribution of the population is approximately normal. We
don’t get to see this distribution so it is again an assumption,
although we can look at the distribution of the sample with
<code>qqnorm(X)</code> and <code>qqnorm(Y)</code>. If we are willing to
assume this, then it follows that the t-statistic follows
t-distribution. What is the p-value under the t-distribution
approximation? Hint: use the <code>t.test</code> function.</li>
<li>With the CLT distribution, we obtained a p-value smaller than 0.05
and with the t-distribution, one that is larger. They can’t both be
right. What best describes the difference?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>A sample size of 12 is not large enough, so we have to use the
t-distribution approximation.<br>
</li>
<li>These are two different assumptions. The t-distribution accounts for
the variability introduced by the estimation of the standard error and
thus, under the null, large values are more probable under the null
distribution.<br>
</li>
<li>The population data is probably not normally distributed so the
t-distribution approximation is wrong.<br>
</li>
<li>Neither assumption is useful. Both are wrong.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Solution to Exercise 3
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>.</li>
<li>.</li>
<li>.</li>
<li>.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-inference-ttests-practice"><p>Content from <a href="inference-ttests-practice.html">t-tests in practice</a></p>
<hr>
<p>Last updated on 2025-08-17 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-ttests-practice.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How are t-tests used in practice?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<hr>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="t-tests-in-practice">t-tests in Practice<a class="anchor" aria-label="anchor" href="#t-tests-in-practice"></a>
</h2>
<hr class="half-width">
<div class="section level4">
<h4 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h4>
<p>We will now demonstrate how to obtain a p-value in practice. We begin
by loading experimental data and walking you through the steps used to
form a t-statistic and compute a p-value. We can perform this task with
just a few lines of code (go to the end of this section to see them).
However, to understand the concepts, we will construct a t-statistic
from “scratch”.</p>
</div>
<div class="section level4">
<h4 id="read-in-and-prepare-data">Read in and prepare data<a class="anchor" aria-label="anchor" href="#read-in-and-prepare-data"></a>
</h4>
<p>We start by reading in the data. A first important step is to
identify which rows are associated with treatment and control, and to
compute the difference in mean.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fWeights</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleMiceWeights.csv"</span><span class="op">)</span> <span class="co"># we read this data in earlier</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">unlist</span></span>
<span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">unlist</span></span>
<span><span class="va">diff</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">diff</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.020833</code></pre>
</div>
<p>We are asked to report a p-value. What do we do? We learned that
<code>diff</code>, referred to as the <em>observed effect size</em>, is
a random variable. We also learned that under the null hypothesis, the
mean of the distribution of <code>diff</code> is 0. What about the
standard error? We also learned that the standard error of this random
variable is the population standard deviation divided by the square root
of the sample size:</p>
<figure><img src="fig/02-stderr.png" class="figure mx-auto d-block"></figure><p>We use the sample standard deviation as an estimate of the population
standard deviation. In R, we simply use the <code>sd</code> function and
the SE is:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sd</span><span class="op">(</span><span class="va">control</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">control</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.8725323</code></pre>
</div>
<p>This is the SE of the sample average, but we actually want the SE of
<code>diff</code>. We saw how statistical theory tells us that the
variance of the difference of two random variables is the sum of its
variances, so we compute the variance and take the square root:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">sqrt</span><span class="op">(</span> </span>
<span>  <span class="fu">var</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">var</span><span class="op">(</span><span class="va">control</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">control</span><span class="op">)</span> </span>
<span>  <span class="op">)</span></span></code></pre>
</div>
<p>Statistical theory tells us that if we divide a random variable by
its SE, we get a new random variable with an SE of 1.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tstat</span> <span class="op">&lt;-</span> <span class="va">diff</span><span class="op">/</span><span class="va">se</span> </span></code></pre>
</div>
<p>This ratio is what we call the t-statistic. It’s the ratio of two
random variables and thus a random variable. Once we know the
distribution of this random variable, we can then easily compute a
p-value.</p>
<p>As explained in the previous section, the CLT tells us that for large
sample sizes, both sample averages <code>mean(treatment)</code> and
<code>mean(control)</code> are normal. Statistical theory tells us that
the difference of two normally distributed random variables is again
normal, so CLT tells us that <code>tstat</code> is approximately normal
with mean 0 (the null hypothesis) and SD 1 (we divided by its SE).</p>
<p>So now to calculate a p-value all we need to do is ask: how often
does a normally distributed random variable exceed <code>diff</code>? R
has a built-in function, <code>pnorm</code>, to answer this specific
question. <code>pnorm(a)</code> returns the probability that a random
variable following the standard normal distribution falls below
<code>a</code>. To obtain the probability that it is larger than
<code>a</code>, we simply use <code>1-pnorm(a)</code>. We want to know
the probability of seeing something as extreme as <code>diff</code>:
either smaller (more negative) than <code>-abs(diff)</code> or larger
than <code>abs(diff)</code>. We call these two regions “tails” and
calculate their size:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">righttail</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu">pnorm</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">tstat</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">lefttail</span> <span class="op">&lt;-</span> <span class="fu">pnorm</span><span class="op">(</span><span class="op">-</span><span class="fu">abs</span><span class="op">(</span><span class="va">tstat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pval</span> <span class="op">&lt;-</span> <span class="va">lefttail</span> <span class="op">+</span> <span class="va">righttail</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">pval</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.0398622</code></pre>
</div>
<p>In this case, the p-value is smaller than 0.05 and using the
conventional cutoff of 0.05, we would call the difference
<em>statistically significant</em>.</p>
<p>Now there is a problem. CLT works for large samples, but is 12 large
enough? A rule of thumb for CLT is that 30 is a large enough sample size
(but this is just a rule of thumb). The p-value we computed is only a
valid approximation if the assumptions hold, which do not seem to be the
case here. However, there is another option other than using CLT.</p>
<p><a name="smallsample"></a></p>
</div>
</section><section><h2 class="section-heading" id="the-t-distribution-in-practice">The t-distribution in Practice<a class="anchor" aria-label="anchor" href="#the-t-distribution-in-practice"></a>
</h2>
<hr class="half-width">
<p>As described earlier, statistical theory offers another useful
result. If the distribution of the population is normal, then we can
work out the exact distribution of the t-statistic without the need for
the CLT. This is a big “if” given that, with small samples, it is hard
to check if the population is normal. But for something like weight, we
suspect that the population distribution is likely well approximated by
normal and that we can use this approximation. Furthermore, we can look
at a qq-plot for the samples. This shows that the approximation is at
least close:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">treatment</span>, col<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">control</span>, col<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-ttests-practice-rendered-data_qqplot-1.png" alt="Quantile-quantile plots for sample against theoretical normal distribution." class="figure mx-auto d-block"><figcaption>
Quantile-quantile plots for sample against theoretical normal
distribution.
</figcaption></figure><p>If we use this approximation, then statistical theory tells us that
the distribution of the random variable <code>tstat</code> follows a
t-distribution. This is a much more complicated distribution than the
normal. The t-distribution has a location parameter like the normal and
another parameter called <em>degrees of freedom</em>. R has a nice
function that actually computes everything for us.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Welch Two Sample t-test

data:  treatment and control
t = 2.0552, df = 20.236, p-value = 0.053
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -0.04296563  6.08463229
sample estimates:
mean of x mean of y
 26.83417  23.81333 </code></pre>
</div>
<p>To see just the p-value, we can use the <code>$</code> extractor:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>,<span class="va">control</span><span class="op">)</span></span>
<span><span class="va">result</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.05299888</code></pre>
</div>
<p>The p-value is slightly bigger now. This is to be expected because
our CLT approximation considered the denominator of <code>tstat</code>
practically fixed (with large samples it practically is), while the
t-distribution approximation takes into account that the denominator
(the standard error of the difference) is a random variable. The smaller
the sample size, the more the denominator varies.</p>
<p>It may be confusing that one approximation gave us one p-value and
another gave us another, because we expect there to be just one answer.
However, this is not uncommon in data analysis. We used different
assumptions, different approximations, and therefore we obtained
different results.</p>
<p>Later, in the power calculation section, we will describe type I and
type II errors. As a preview, we will point out that the test based on
the CLT approximation is more likely to incorrectly reject the null
hypothesis (a false positive), while the t-distribution is more likely
to incorrectly accept the null hypothesis (false negative).</p>
<div class="section level4">
<h4 id="running-the-t-test-in-practice">Running the t-test in practice<a class="anchor" aria-label="anchor" href="#running-the-t-test-in-practice"></a>
</h4>
<p>Now that we have gone over the concepts, we can show the relatively
simple code that one would use to actually compute a t-test:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">unlist</span></span>
<span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">unlist</span></span>
<span><span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Welch Two Sample t-test

data:  treatment and control
t = 2.0552, df = 20.236, p-value = 0.053
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -0.04296563  6.08463229
sample estimates:
mean of x mean of y
 26.83417  23.81333 </code></pre>
</div>
<p>The arguments to <code>t.test</code> can be of type
<em>data.frame</em> and thus we do not need to unlist them into numeric
objects.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>.</li>
<li>.</li>
<li>.</li>
<li>.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-inference-ci"><p>Content from <a href="inference-ci.html">Confidence Intervals</a></p>
<hr>
<p>Last updated on 2025-08-17 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-ci.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is a confidence interval?</li>
<li>When is it best to use a confidence interval?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Define a confidence interval.</li>
<li>Explain why confidence intervals offer more information than
p-values alone.</li>
<li>Calculate a confidence interval.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="confidence-intervals">Confidence Intervals<a class="anchor" aria-label="anchor" href="#confidence-intervals"></a>
</h2>
<hr class="half-width">
<p>We have described how to compute p-values which are ubiquitous in the
life sciences. However, we do not recommend reporting p-values as the
only statistical summary of your results. The reason is simple:
statistical significance does not guarantee scientific significance.
With large enough sample sizes, one might detect a statistically
significance difference in weight of, say, 1 microgram. But is this an
important finding? Would we say a diet results in higher weight if the
increase is less than a fraction of a percent? The problem with
reporting only p-values is that you will not provide a very important
piece of information: the effect size. Recall that the effect size is
the observed difference. Sometimes the effect size is divided by the
mean of the control group and so expressed as a percent increase.</p>
<p>A much more attractive alternative is to report confidence intervals.
A confidence interval includes information about your estimated effect
size and the uncertainty associated with this estimate. Here we use the
mice data to illustrate the concept behind confidence intervals.</p>
<div class="section level4">
<h4 id="confidence-interval-for-population-mean">Confidence Interval for Population Mean<a class="anchor" aria-label="anchor" href="#confidence-interval-for-population-mean"></a>
</h4>
<p>Before we show how to construct a confidence interval for the
difference between the two groups, we will show how to construct a
confidence interval for the population mean of control female mice. Then
we will return to the group difference after we’ve learned how to build
confidence intervals in the simple case. We start by reading in the data
and selecting the appropriate rows:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"./data/mice_pheno.csv"</span><span class="op">)</span> <span class="co"># we read this in earlier</span></span>
<span><span class="va">chowPopulation</span> <span class="op">&lt;-</span> <span class="va">pheno</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">Sex</span><span class="op">==</span><span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span><span class="op">==</span><span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="va">unlist</span></span></code></pre>
</div>
<p>The population average <i>μ<sub>X</sub></i> is our parameter of
interest here:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mu_chow</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">mu_chow</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23.89338</code></pre>
</div>
<p>We are interested in estimating this parameter. In practice, we do
not get to see the entire population so, as we did for p-values, we
demonstrate how we can use samples to do this. Let’s start with a sample
of size 30:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span><span class="va">chow</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">chowPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 24.03267</code></pre>
</div>
<p>We know this is a random variable, so the sample average will not be
a perfect estimate. In fact, because in this illustrative example we
know the value of the parameter, we can see that they are not exactly
the same. A confidence interval is a statistical way of reporting our
finding, the sample average, in a way that explicitly summarizes the
variability of our random variable.</p>
<p>With a sample size of 30, we will use the CLT. The CLT tells us that
X̄ or <code>mean(chow)</code> follows a normal distribution with mean
<i>μ<sub>X</sub></i> or <code>mean(chowPopulation)</code> and standard
error approximately <i>s<sub>X</sub> / √N</i> or:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">sd</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="va">N</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.6875646</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="defining-the-interval">Defining the Interval<a class="anchor" aria-label="anchor" href="#defining-the-interval"></a>
</h4>
<p>A 95% confidence interval (we can use percentages other than 95%) is
a random interval with a 95% probability of falling on the parameter we
are estimating. Keep in mind that saying 95% of random intervals will
fall on the true value (our definition above) is <em>not the same</em>
as saying there is a 95% chance that the true value falls in our
interval. To construct it, we note that the CLT tells us that <i>√N</i>
(X̄ <i>- μ<sub>X</sub>) / s<sub>X</sub></i> follows a normal distribution
with mean 0 and SD 1. This implies that the probability of this
event:</p>
<figure><img src="fig/02-confint.png" class="figure mx-auto d-block"></figure><p>which written in R code is:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pnorm</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span> <span class="op">-</span> <span class="fu">pnorm</span><span class="op">(</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.9544997</code></pre>
</div>
<p>…is about 95% (to get closer use <code>qnorm(1 - 0.05/2)</code>
instead of 2). Now do some basic algebra to clear out everything and
leave <i>μ<sub>X</sub></i> alone in the middle and you get that the
following event:</p>
<figure><img src="fig/02-event-confint.png" class="figure mx-auto d-block"></figure><p>has a probability of 95%.</p>
<p>Be aware that it is the edges of the interval X̄ <i>± 2s<sub>X</sub> /
√N</i>, not <i>μ<sub>X</sub></i>, that are random. Again, the definition
of the confidence interval is that 95% of <em>random intervals</em> will
contain the true, fixed value <i>μ<sub>X</sub></i>. For a specific
interval that has been calculated, the probability is either 0 or 1 that
it contains the fixed population mean <i>μ<sub>X</sub></i>.</p>
<p>Let’s demonstrate this logic through simulation. We can construct
this interval with R relatively easily:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Q</span> <span class="op">&lt;-</span> <span class="fu">qnorm</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">interval</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">-</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">+</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span> <span class="op">)</span></span>
<span><span class="va">interval</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 22.68506 25.38027</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interval</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="va">mu_chow</span> <span class="op">&amp;</span> <span class="va">interval</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">mu_chow</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<p>which happens to cover <i>μ<sub>X</sub></i> or
<code>mean(chowPopulation)</code>. However, we can take another sample
and we might not be as lucky. In fact, the theory tells us that we will
cover <i>μ<sub>X</sub></i> 95% of the time. Because we have access to
the population data, we can confirm this by taking several new
samples:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">250</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">+</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">7</span>,<span class="fl">7</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"n"</span>,</span>
<span>     xlab<span class="op">=</span><span class="st">"weight"</span>, ylab<span class="op">=</span><span class="st">"interval"</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="va">B</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">B</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">chow</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">chowPopulation</span>,<span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">sd</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">interval</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">-</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">+</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span><span class="op">)</span></span>
<span>  <span class="va">covered</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">interval</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&amp;</span> <span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">interval</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">color</span> <span class="op">&lt;-</span> <span class="fu">ifelse</span><span class="op">(</span><span class="va">covered</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="fu">lines</span><span class="op">(</span><span class="va">interval</span>, <span class="fu">c</span><span class="op">(</span><span class="va">i</span>,<span class="va">i</span><span class="op">)</span>,col<span class="op">=</span><span class="va">color</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-ci-rendered-confidence_interval_n30-1.png" alt="We show 250 random realizations of 95% confidence intervals. The color denotes if the interval fell on the parameter or not." class="figure mx-auto d-block"><figcaption>
We show 250 random realizations of 95% confidence intervals. The color
denotes if the interval fell on the parameter or not.
</figcaption></figure><p>You can run this repeatedly to see what happens. You will see that in
about 5% of the cases, we fail to cover <i>μ<sub>X</sub></i>.</p>
</div>
<div class="section level4">
<h4 id="small-sample-size-and-the-clt">Small Sample Size and the CLT<a class="anchor" aria-label="anchor" href="#small-sample-size-and-the-clt"></a>
</h4>
<p>For <em>N = 30</em>, the CLT works very well. However, if <em>N =
5</em>, do these confidence intervals work as well? We used the CLT to
create our intervals, and with <em>N = 5</em> it may not be as useful an
approximation. We can confirm this with a simulation:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">+</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">7</span>,<span class="fl">7</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"n"</span>,</span>
<span>     xlab<span class="op">=</span><span class="st">"weight"</span>, ylab<span class="op">=</span><span class="st">"interval"</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="va">B</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Q</span> <span class="op">&lt;-</span> <span class="fu">qnorm</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">B</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">chow</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">chowPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">sd</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">interval</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">-</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">+</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span><span class="op">)</span></span>
<span>  <span class="va">covered</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">interval</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&amp;</span> <span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">interval</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">color</span> <span class="op">&lt;-</span> <span class="fu">ifelse</span><span class="op">(</span><span class="va">covered</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="fu">lines</span><span class="op">(</span><span class="va">interval</span>, <span class="fu">c</span><span class="op">(</span><span class="va">i</span>,<span class="va">i</span><span class="op">)</span>,col<span class="op">=</span><span class="va">color</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-ci-rendered-confidence_interval_n5-1.png" alt="We show 250 random realizations of 95% confidence intervals, but now for a smaller sample size. The confidence interval is based on the CLT approximation. The color denotes if the interval fell on the parameter or not." class="figure mx-auto d-block"><figcaption>
We show 250 random realizations of 95% confidence intervals, but now for
a smaller sample size. The confidence interval is based on the CLT
approximation. The color denotes if the interval fell on the parameter
or not.
</figcaption></figure><p>Despite the intervals being larger (we are dividing by <i>√5</i>
instead of <i>√30</i> ), we see many more intervals not covering
<i>μ<sub>X</sub></i>. This is because the CLT is incorrectly telling us
that the distribution of the <code>mean(chow)</code> is approximately
normal with standard deviation 1 when, in fact, it has a larger standard
deviation and a fatter tail (the parts of the distribution going to <i>±
∞</i>). This mistake affects us in the calculation of <code>Q</code>,
which assumes a normal distribution and uses <code>qnorm</code>. The
t-distribution might be more appropriate. All we have to do is re-run
the above, but change how we calculate <code>Q</code> to use
<code>qt</code> instead of <code>qnorm</code>.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">+</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">7</span>,<span class="fl">7</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"n"</span>,</span>
<span>     xlab<span class="op">=</span><span class="st">"weight"</span>, ylab<span class="op">=</span><span class="st">"interval"</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="va">B</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##Q &lt;- qnorm(1- 0.05/2) ##no longer normal so use:</span></span>
<span><span class="va">Q</span> <span class="op">&lt;-</span> <span class="fu">qt</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">2</span>, df<span class="op">=</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">B</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">chow</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">chowPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">sd</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">interval</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">-</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">chow</span><span class="op">)</span> <span class="op">+</span> <span class="va">Q</span> <span class="op">*</span> <span class="va">se</span> <span class="op">)</span></span>
<span>  <span class="va">covered</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">interval</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&amp;</span> <span class="fu">mean</span><span class="op">(</span><span class="va">chowPopulation</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">interval</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">color</span> <span class="op">&lt;-</span> <span class="fu">ifelse</span><span class="op">(</span><span class="va">covered</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="fu">lines</span><span class="op">(</span><span class="va">interval</span>, <span class="fu">c</span><span class="op">(</span><span class="va">i</span>,<span class="va">i</span><span class="op">)</span>,col<span class="op">=</span><span class="va">color</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-ci-rendered-confidence_interval_tdist_n5-1.png" alt="We show 250 random realizations of 95% confidence intervals, but now for a smaller sample size. The confidence is now based on the t-distribution approximation. The color denotes if the interval fell on the parameter or not." class="figure mx-auto d-block"><figcaption>
We show 250 random realizations of 95% confidence intervals, but now for
a smaller sample size. The confidence is now based on the t-distribution
approximation. The color denotes if the interval fell on the parameter
or not.
</figcaption></figure><p>Now the intervals are made bigger. This is because the t-distribution
has fatter tails and therefore:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qt</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">2</span>, df<span class="op">=</span><span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2.776445</code></pre>
</div>
<p>is bigger than…</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qnorm</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.959964</code></pre>
</div>
<p>…which makes the intervals larger and hence cover
<i>μ<sub>X</sub></i> more frequently; in fact, about 95% of the
time.</p>
</div>
<div class="section level4">
<h4 id="connection-between-confidence-intervals-and-p-values">Connection Between Confidence Intervals and p-values<a class="anchor" aria-label="anchor" href="#connection-between-confidence-intervals-and-p-values"></a>
</h4>
<p>We recommend that in practice confidence intervals be reported
instead of p-values. If for some reason you are required to provide
p-values, or required that your results are significant at the 0.05 of
0.01 levels, confidence intervals do provide this information.</p>
<p>If we are talking about a t-test p-value, we are asking if
differences as extreme as the one we observe, Ȳ - X̄, are likely when the
difference between the population averages is actually equal to zero. So
we can form a confidence interval with the observed difference. Instead
of writing Ȳ - X̄ repeatedly, let’s define this difference as a new
variable <i>d ≡</i> Ȳ - X̄.</p>
<p>Suppose you use CLT and report</p>
<p><img src="fig/d-plusmn.png" class="figure"> with</p>
<figure><img src="fig/s-sub-d.png" class="figure mx-auto d-block"></figure><p>as a 95% confidence interval for the difference and this interval
does not include 0 (a false positive). Because the interval does not
include 0, this implies that either</p>
<p><img src="fig/bigD-gt-zero.png" class="figure"> or</p>
<p><img src="fig/smalld-lt-zero.png" class="figure"> This suggests that either</p>
<p><img src="fig/sqrtNd-gt-2.png" class="figure"> or</p>
<p><img src="fig/sqrtNd-lt-2.png" class="figure"> This then implies that the
t-statistic is more extreme than 2, which in turn suggests that the
p-value must be smaller than 0.05 (approximately, for a more exact
calculation use <code>qnorm(.05/2)</code> instead of 2). The same
calculation can be made if we use the t-distribution instead of CLT
(with <code>qt(.05/2, df = 2 * N-2)</code>). In summary, if a 95% or 99%
confidence interval does not include 0, then the p-value must be smaller
than 0.05 or 0.01 respectively.</p>
<p>Note that the confidence interval for the difference <em>d</em> is
provided by the <code>t.test</code> function:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">controlIndex</span> <span class="op">&lt;-</span> <span class="fu">which</span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">Diet</span><span class="op">==</span><span class="st">"chow"</span><span class="op">)</span></span>
<span><span class="va">treatmentIndex</span> <span class="op">&lt;-</span> <span class="fu">which</span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">Diet</span><span class="op">==</span><span class="st">"hf"</span><span class="op">)</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">[</span><span class="va">controlIndex</span>, <span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">[</span><span class="va">treatmentIndex</span>, <span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">conf.int</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2.231533 3.906857
attr(,"conf.level")
[1] 0.95</code></pre>
</div>
<p>In this case, the 95% confidence interval does include 0 and we
observe that the p-value is larger than 0.05 as predicted. If we change
this to a 90% confidence interval, then:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span>, conf.level<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span><span class="op">$</span><span class="va">conf.int</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2.366479 3.771911
attr(,"conf.level")
[1] 0.9</code></pre>
</div>
<p>0 is no longer in the confidence interval (which is expected because
the p-value is smaller than 0.10).</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>A confidence intervals is a better alternative to reporting p-values
alone.</li>
<li>A confidence interval estimates effect size and the uncertainty
associated with this estimate.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-inference-power-calc"><p>Content from <a href="inference-power-calc.html">Power Calculations</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-power-calc.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is statistical power?</li>
<li>How is power calculated?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li><ul><li>
</li></ul></li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="power-calculations">Power Calculations<a class="anchor" aria-label="anchor" href="#power-calculations"></a>
</h2>
<hr class="half-width">
<div class="section level4">
<h4 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h4>
<p>We have used the example of the effects of two different diets on the
weight of mice. Since in this illustrative example we have access to the
population, we know that in fact there is a substantial (about 10%)
difference between the average weights of the two populations:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"./data/mice_pheno.csv"</span><span class="op">)</span> <span class="co"># Previously downloaded </span></span>
<span></span>
<span><span class="va">controlPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span></span>
<span><span class="va">hfPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span></span>
<span><span class="va">mu_hf</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">hfPopulation</span><span class="op">)</span></span>
<span><span class="va">mu_control</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">mu_hf</span> <span class="op">-</span> <span class="va">mu_control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2.375517</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">print</span><span class="op">(</span><span class="op">(</span><span class="va">mu_hf</span> <span class="op">-</span> <span class="va">mu_control</span><span class="op">)</span><span class="op">/</span><span class="va">mu_control</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span> <span class="co">#percent increase</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 9.942157</code></pre>
</div>
<p>We have also seen that, in some cases, when we take a sample and
perform a t-test, we don’t always get a p-value smaller than 0.05. For
example, here is a case where we take a sample of 5 mice and don’t
achieve statistical significance at the 0.05 level:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">hf</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">hfPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="fu">t.test</span><span class="op">(</span><span class="va">hf</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.5806661</code></pre>
</div>
<p>Did we make a mistake? By not rejecting the null hypothesis, are we
saying the diet has no effect? The answer to this question is no. All we
can say is that we did not reject the null hypothesis. But this does not
necessarily imply that the null is true. The problem is that, in this
particular instance, we don’t have enough <em>power</em>, a term we are
now going to define. If you are doing scientific research, it is very
likely that you will have to do a power calculation at some point. In
many cases, it is an ethical obligation as it can help you avoid
sacrificing mice unnecessarily or limiting the number of human subjects
exposed to potential risk in a study. Here we explain what statistical
power means.</p>
</div>
<div class="section level4">
<h4 id="types-of-error">Types of Error<a class="anchor" aria-label="anchor" href="#types-of-error"></a>
</h4>
<p>Whenever we perform a statistical test, we are aware that we may make
a mistake. This is why our p-values are not 0. Under the null, there is
always a positive, perhaps very small, but still positive chance that we
will reject the null when it is true. If the p-value is 0.05, it will
happen 1 out of 20 times. This <em>error</em> is called <em>type I
error</em> by statisticians.</p>
<figure><img src="fig/null-hypothesis.png" alt="Null distribution showing type I error as alpha" class="figure mx-auto d-block"></figure><p>A type I error is defined as rejecting the null when we should not.
This is also referred to as a false positive. So why do we then use
0.05? Shouldn’t we use 0.000001 to be really sure? The reason we don’t
use infinitesimal cut-offs to avoid type I errors at all cost is that
there is another error we can commit: to not reject the null when we
should. This is called a <em>type II error</em> or a false negative.</p>
<figure><img src="fig/alternative-hypothesis.png" alt="Alternative hypothesis showing type II error as beta" class="figure mx-auto d-block"></figure><p>The R code analysis above shows an example of a false negative: we
did not reject the null hypothesis (at the 0.05 level) and, because we
happen to know and peeked at the true population means, we know there is
in fact a difference. Had we used a p-value cutoff of 0.25, we would not
have made this mistake. However, in general, are we comfortable with a
type I error rate of 1 in 4? Usually we are not.</p>
</div>
<div class="section level4">
<h4 id="the-0-05-and-0-01-cut-offs-are-arbitrary">The 0.05 and 0.01 Cut-offs Are Arbitrary<a class="anchor" aria-label="anchor" href="#the-0-05-and-0-01-cut-offs-are-arbitrary"></a>
</h4>
<p>Most journals and regulatory agencies frequently insist that results
be significant at the 0.01 or 0.05 levels. Of course there is nothing
special about these numbers other than the fact that some of the first
papers on p-values used these values as examples. Part of the goal of
this book is to give readers a good understanding of what p-values and
confidence intervals are so that these choices can be judged in an
informed way. Unfortunately, in science, these cut-offs are applied
somewhat mindlessly, but that topic is part of a complicated debate.</p>
</div>
<div class="section level4">
<h4 id="power-calculation">Power Calculation<a class="anchor" aria-label="anchor" href="#power-calculation"></a>
</h4>
<p>Power is the probability of rejecting the null when the null is
false. Of course “when the null is false” is a complicated statement
because it can be false in many ways.</p>
<figure><img src="fig/delta-eq.png" class="figure mx-auto d-block"></figure><p>could be anything and the power actually depends on this parameter.
It also depends on the standard error of your estimates which in turn
depends on the sample size and the population standard deviations. In
practice, we don’t know these so we usually report power for several
plausible values of <i>Δ, σ<sub>X</sub>, σ<sub>Y</sub></i> and various
sample sizes. Statistical theory gives us formulas to calculate power.
The <code>pwr</code> package performs these calculations for you. Here
we will illustrate the concepts behind power by coding up simulations in
R.</p>
<p>Suppose our sample size is:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span></code></pre>
</div>
<p>and we will reject the null hypothesis at:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span></code></pre>
</div>
<p>What is our power with this particular data? We will compute this
probability by re-running the exercise many times and calculating the
proportion of times the null hypothesis is rejected. Specifically, we
will run:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">2000</span></span></code></pre>
</div>
<p>simulations. The simulation is as follows: we take a sample of size
<em>N</em> from both control and treatment groups, we perform a t-test
comparing these two, and report if the p-value is less than
<code>alpha</code> or not. We write a function that does this:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reject</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">N</span>, <span class="va">alpha</span><span class="op">=</span><span class="fl">0.05</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="va">hf</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">hfPopulation</span>, <span class="va">N</span><span class="op">)</span> </span>
<span>   <span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span>   <span class="va">pval</span> <span class="op">&lt;-</span> <span class="fu">t.test</span><span class="op">(</span><span class="va">hf</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>   <span class="va">pval</span> <span class="op">&lt;</span> <span class="va">alpha</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Here is an example of one simulation for a sample size of 12. The
call to <code>reject</code> answers the question “Did we reject?”</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reject</span><span class="op">(</span><span class="fl">12</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<p>Now we can use the <code>replicate</code> function to do this
<code>B</code> times.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rejections</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="fu">reject</span><span class="op">(</span><span class="va">N</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Our power is just the proportion of times we correctly reject. So
with <em>N = 12</em> our power is only:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mean</span><span class="op">(</span><span class="va">rejections</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.2155</code></pre>
</div>
<p>This explains why the t-test was not rejecting when we knew the null
was false. With a sample size of just 12, our power is about 23%. To
guard against false positives at the 0.05 level, we had set the
threshold at a high enough level that resulted in many type II
errors.</p>
<p>Let’s see how power improves with N. We will use the function
<code>sapply</code>, which applies a function to each of the elements of
a vector. We want to repeat the above for the following sample size:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ns</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">5</span>, <span class="fl">50</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<p>So we use <code>apply</code> like this:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">power</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">Ns</span>,<span class="kw">function</span><span class="op">(</span><span class="va">N</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">rejections</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="fu">reject</span><span class="op">(</span><span class="va">N</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">mean</span><span class="op">(</span><span class="va">rejections</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>For each of the three simulations, the above code returns the
proportion of times we reject. Not surprisingly power increases with
N:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">Ns</span>, <span class="va">power</span>, type<span class="op">=</span><span class="st">"b"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-power-calc-rendered-power_versus_sample_size-1.png" alt="Power plotted against sample size." class="figure mx-auto d-block"><figcaption>
Power plotted against sample size.
</figcaption></figure><p>Similarly, if we change the level <code>alpha</code> at which we
reject, power changes. The smaller I want the chance of type I error to
be, the less power I will have. Another way of saying this is that we
trade off between the two types of error. We can see this by writing
similar code, but keeping <em>N</em> fixed and considering several
values of <code>alpha</code>:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span><span class="va">alphas</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.05</span>, <span class="fl">0.01</span>, <span class="fl">0.001</span>, <span class="fl">0.0001</span><span class="op">)</span></span>
<span><span class="va">power</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">alphas</span>, <span class="kw">function</span><span class="op">(</span><span class="va">alpha</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">rejections</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="fu">reject</span><span class="op">(</span><span class="va">N</span>, alpha<span class="op">=</span><span class="va">alpha</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">mean</span><span class="op">(</span><span class="va">rejections</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">alphas</span>, <span class="va">power</span>, xlab<span class="op">=</span><span class="st">"alpha"</span>, type<span class="op">=</span><span class="st">"b"</span>, log<span class="op">=</span><span class="st">"x"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-power-calc-rendered-power_versus_alpha-1.png" alt="Power plotted against cut-off." class="figure mx-auto d-block"><figcaption>
Power plotted against cut-off.
</figcaption></figure><p>Note that the x-axis in this last plot is in the log scale.</p>
<p>There is no “right” power or “right” alpha level, but it is important
that you understand what each means.</p>
<p>To see this clearly, you could create a plot with curves of power
versus N. Show several curves in the same plot with color representing
alpha level.</p>
</div>
<div class="section level4">
<h4 id="p-values-are-arbitrary-under-the-alternative-hypothesis">p-values Are Arbitrary under the Alternative Hypothesis<a class="anchor" aria-label="anchor" href="#p-values-are-arbitrary-under-the-alternative-hypothesis"></a>
</h4>
<p>Another consequence of what we have learned about power is that
p-values are somewhat arbitrary when the null hypothesis is not true and
therefore the <em>alternative</em> hypothesis is true (the difference
between the population means is not zero). When the alternative
hypothesis is true, we can make a p-value as small as we want simply by
increasing the sample size (supposing that we have an infinite
population to sample from). We can show this property of p-values by
drawing larger and larger samples from our population and calculating
p-values. This works because, in our case, we know that the alternative
hypothesis is true, since we have access to the populations and can
calculate the difference in their means.</p>
<p>First write a function that returns a p-value for a given sample size
<em>N</em>:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">calculatePvalue</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">N</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">hf</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">hfPopulation</span>,<span class="va">N</span><span class="op">)</span> </span>
<span>   <span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>,<span class="va">N</span><span class="op">)</span></span>
<span>   <span class="fu">t.test</span><span class="op">(</span><span class="va">hf</span>,<span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>We have a limit here of 200 for the high-fat diet population, but we
can see the effect well before we get to 200. For each sample size, we
will calculate a few p-values. We can do this by repeating each value of
<em>N</em> a few times.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ns</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">10</span>,<span class="fl">200</span>,by<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">Ns_rep</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="va">Ns</span>, each<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<p>Again we use <code>sapply</code> to run our simulations:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pvalues</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">Ns_rep</span>, <span class="va">calculatePvalue</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can plot the 10 p-values we generated for each sample
size:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">Ns_rep</span>, <span class="va">pvalues</span>, log<span class="op">=</span><span class="st">"y"</span>, xlab<span class="op">=</span><span class="st">"sample size"</span>,</span>
<span>     ylab<span class="op">=</span><span class="st">"p-values"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">.01</span>, <span class="fl">.05</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"red"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-power-calc-rendered-pvals_decrease-1.png" alt="p-values from random samples at varying sample size. The actual value of the p-values decreases as we increase sample size whenever the alternative hypothesis is true." class="figure mx-auto d-block"><figcaption>
p-values from random samples at varying sample size. The actual value of
the p-values decreases as we increase sample size whenever the
alternative hypothesis is true.
</figcaption></figure><p>Note that the y-axis is log scale and that the p-values show a
decreasing trend all the way to 10<sup>-8</sup> as the sample size gets
larger. The standard cutoffs of 0.01 and 0.05 are indicated with
horizontal red lines.</p>
<p>It is important to remember that p-values are not more interesting as
they become very very small. Once we have convinced ourselves to reject
the null hypothesis at a threshold we find reasonable, having an even
smaller p-value just means that we sampled more mice than was necessary.
Having a larger sample size does help to increase the precision of our
estimate of the difference <i>Δ</i>, but the fact that the p-value
becomes very very small is just a natural consequence of the mathematics
of the test. The p-values get smaller and smaller with increasing sample
size because the numerator of the t-statistic has <i>√N</i> (for equal
sized groups, and a similar effect occurs when <i>M ≠ N</i>). Therefore,
if <i>Δ</i> is non-zero, the t-statistic will increase with
<i>N</i>.</p>
<p>Therefore, a better statistic to report is the effect size with a
confidence interval or some statistic which gives the reader a sense of
the change in a meaningful scale. We can report the effect size as a
percent by dividing the difference and the confidence interval by the
control population mean:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">hf</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">hfPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="va">diff</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">hf</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span>
<span><span class="va">diff</span> <span class="op">/</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 5.196293</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">t.test</span><span class="op">(</span><span class="va">hf</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">conf.int</span> <span class="op">/</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] -7.775307 18.167892
attr(,"conf.level")
[1] 0.95</code></pre>
</div>
<p>In addition, we can report a statistic called <a href="https://en.wikipedia.org/wiki/Effect_size#Cohen.27s_d" class="external-link">Cohen’s
d</a>, which is the difference between the groups divided by the pooled
standard deviation of the two groups.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sd_pool</span> <span class="op">&lt;-</span> <span class="fu">sqrt</span><span class="op">(</span><span class="op">(</span><span class="op">(</span><span class="va">N</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu">var</span><span class="op">(</span><span class="va">hf</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="va">N</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fu">var</span><span class="op">(</span><span class="va">control</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">N</span> <span class="op">-</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">diff</span> <span class="op">/</span> <span class="va">sd_pool</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.3397886</code></pre>
</div>
<p>This tells us how many standard deviations of the data the mean of
the high-fat diet group is from the control group. Under the alternative
hypothesis, unlike the t-statistic which is guaranteed to increase, the
effect size and Cohen’s d will become more precise.</p>
<div id="exercises" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercises" class="callout-inner">
<h3 class="callout-title">Exercises</h3>
<div class="callout-content">
<p>For these exercises we will load the babies dataset from babies.txt.
We will use this data to review the concepts behind the p-values and
then test confidence interval concepts.
<code><a href="https://github.com/wch/downloader" class="external-link">library(downloader)</a></code><br><code>url&lt;-"https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/babies.txt"</code><br><code>filename &lt;- basename(url)</code><br><code>download(url, destfile=filename)</code><br><code>babies &lt;- read.table("babies.txt", header=TRUE)</code><br>
This is a large dataset (1,236 cases), and we will pretend that it
contains the entire population in which we are interested. We will study
the differences in birth weight between babies born to smoking and
non-smoking mothers. First, let’s split this into two birth weight
datasets: one of birth weights to non-smoking mothers and the other of
birth weights to smoking mothers.<br><code>bwt.nonsmoke &lt;- filter(babies, smoke==0) %&gt;%</code><br><code>select(bwt) %&gt;%</code><br><code>unlist</code><br><code>bwt.smoke &lt;- filter(babies, smoke==1) %&gt;%</code><br><code>select(bwt) %&gt;%</code><br><code>unlist</code><br>
Now, we can look for the true population difference in means between
smoking and non-smoking birth weights.<br><code>library(rafalib)</code><br><code>mean(bwt.nonsmoke) - mean(bwt.smoke)</code><br><code>popsd(bwt.nonsmoke)</code><br><code>popsd(bwt.smoke)</code><br>
The population difference of mean birth weights is about 8.9 ounces. The
standard deviations of the nonsmoking and smoking groups are about 17.4
and 18.1 ounces, respectively. As we did with the mouse weight data,
this assessment interactively reviews inference concepts using
simulations in R. We will treat the babies dataset as the full
population and draw samples from it to simulate individual experiments.
We will then ask whether somebody who only received the random samples
would be able to draw correct conclusions about the population. We are
interested in testing whether the birth weights of babies born to
non-smoking mothers are significantly different from the birth weights
of babies born to smoking mothers.</p>
<ol style="list-style-type: decimal">
<li>Set the seed at 1 and obtain two samples, each of size <i>N =
25</i>, from non-smoking mothers (<code>dat.ns</code>) and smoking
mothers (<code>dat.s</code>). Compute the t-statistic (call it
<code>tval</code>).</li>
<li>Recall that we summarize our data using a t-statistic because we
know that in situations where the null hypothesis is true (what we mean
when we say “under the null”) and the sample size is relatively large,
this t-value will have an approximate standard normal distribution.
Because we know the distribution of the t-value under the null, we can
quantitatively determine how unusual the observed t-value would be if
the null hypothesis were true. The standard procedure is to examine the
probability a t-statistic that actually does follow the null hypothesis
would have larger absolute value than the absolute value of the t-value
we just observed – this is called a two-sided test. We have computed
these by taking one minus the area under the standard normal curve
between <code>-abs(tval)</code> and <code>abs(tval)</code>. In R, we can
do this by using the <code>pnorm</code> function, which computes the
area under a normal curve from negative infinity up to the value given
as its first argument:</li>
<li>Because of the symmetry of the standard normal distribution, there
is a simpler way to calculate the probability that a t-value under the
null could have a larger absolute value than <code>tval</code>. Choose
the simplified calculation from the following:<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>1 - 2 * pnorm(abs(tval))<br>
</li>
<li>1 - 2 * pnorm(-abs(tval))<br>
</li>
<li>1 - pnorm(-abs(tval))<br>
</li>
<li>2 * pnorm(-abs(tval))</li>
</ol>
<ol start="4" style="list-style-type: decimal">
<li>By reporting only p-values, many scientific publications provide an
incomplete story of their findings. As we have mentioned, with very
large sample sizes, scientifically insignificant differences between two
groups can lead to small p-values. Confidence intervals are more
informative as they include the estimate itself. Our estimate of the
difference between babies of smoker and non-smokers:
<code>mean(dat.s)  -  mean(dat.ns)</code>. If we use the CLT, what
quantity would we add and subtract to this estimate to obtain a 99%
confidence interval?</li>
<li>If instead of CLT, we use the t-distribution approximation, what do
we add and subtract (use 2 * N - 2 degrees of freedom)?</li>
<li>Why are the values from 4 and 5 so similar?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Coincidence.<br>
</li>
<li>They are both related to 99% confidence intervals.<br>
</li>
<li>N and thus the degrees of freedom is large enough to make the normal
and t-distributions very similar.<br>
</li>
<li>They are actually quite different, differing by more than 1
ounce.</li>
</ol>
<ol start="7" style="list-style-type: decimal">
<li>No matter which way you compute it, the p-value <code>pval</code> is
the probability that the null hypothesis could have generated a
t-statistic more extreme than what we observed: <code>tval</code>. If
the p-value is very small, this means that observing a value more
extreme than <code>tval</code> would be very rare if the null hypothesis
were true, and would give strong evidence that we should reject the null
hypothesis. We determine how small the p-value needs to be to reject the
null by deciding how often we would be willing to mistakenly reject the
null hypothesis.<br>
The standard decision rule is the following: choose some small value
<i>α</i> (in most disciplines the conventional choice is <i>α</i> 0:05)
and reject the null hypothesis if the p-value is less than <i>α</i>. We
call the significance level of the test.<br>
It turns out that if we follow this decision rule, the probability that
we will reject the null hypothesis by mistake is equal to <i>α</i>.
(This fact is not immediately obvious and requires some probability
theory to show.) We call the event of rejecting the null hypothesis,
when it is in fact true, a Type I error, we call the probability of
making a Type I error, the Type I error rate, and we say that rejecting
the null hypothesis when the p-value is less than <i>α</i>, controls the
Type I error rate so that it is equal to <i>α</i>. We will see a number
of decision rules that we use in order to control the probabilities of
other types of errors. Often, we will guarantee that the probability of
an error is less than some level, but, in this case, we can guarantee
that the probability of a Type I error is exactly equal to
<i>α</i>.<br>
Which of the following sentences about a Type I error is not true?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>The following is another way to describe a Type I error: you decided
to reject the null hypothesis on the basis of data that was actually
generated by the null hypothesis.<br>
</li>
<li>The following is the another way to describe a Type I error: due to
random fluctuations, even though the data you observed were actually
generated by the null hypothesis, the p-value calculated from the
observed data was small, so you rejected it.<br>
</li>
<li>From the original data alone, you can tell whether you have made a
Type I error.<br>
</li>
<li>In scientific practice, a Type I error constitutes reporting a
“significant” result when there is actually no result.</li>
</ol>
<ol start="8" style="list-style-type: decimal">
<li>In the simulation we have set up here, we know the null hypothesis
is false – the true value of difference in means is actually around 8.9.
Thus, we are concerned with how often the decision rule outlined in the
last section allows us to conclude that the null hypothesis is actually
false. In other words, we would like to quantify the Type II error rate
of the test, or the probability that we fail to reject the null
hypothesis when the alternative hypothesis is true.<br>
Unlike the Type I error rate, which we can characterize by assuming that
the null hypothesis of “no difference” is true, the Type II error rate
cannot be computed by assuming the alternative hypothesis alone because
the alternative hypothesis alone does not specify a particular value for
the difference. It thus does not nail down a specific distribution for
thet-value under the alternative.<br>
For this reason, when we study the Type II error rate of a hypothesis
testing procedure, we need to assume a particular effect size, or
hypothetical size of the difference between population means, that we
wish to target. We ask questions such as “what is the smallest
difference I could reliably distinguish from 0 given my sample size N?”
or, more commonly,“How big does N have to be in order to detect that the
absolute value of the difference is greater than zero?” Type II error
control plays a major role in designing data collection procedures
before you actually see the data, so that you know the test you will run
has enough sensitivity or power. Power is one minus the Type II error
rate, or the probability that you will reject the null hypothesis when
the alternative hypothesis is true.<br>
There are several aspects of a hypothesis test that affect its power for
a particular effect size. Intuitively, setting a lower <i>α</i>
decreases the power of the test for a given effect size because the null
hypothesis will be more difficult to reject. This means that for an
experiment with fixed parameters (i.e., with a predetermined sample
size, recording mechanism, etc), the power of the hypothesis test trades
off with its Type I error rate, no matter what effect size you
target.<br>
We can explore the trade off of power and Type I error concretely using
the babies data. Since we have the full population, we know what the
true effect size is (about 8.93) and we can compute the power of the
test for true difference between populations.<br>
Set the seed at 1 and take a random sample of N = 5 measurements from
each of the smoking and nonsmoking datasets. What is the p-value (use
the <code>t-test</code> function)?</li>
<li>The p-value is larger than 0.05 so using the typical cut-off, we
would not reject. This is a type II error. Which of the following is not
a way to decrease this type of error?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Increase our chance of a type I error.<br>
</li>
<li>Take a larger sample size<br>
</li>
<li>Find a population for which the null is not true.<br>
</li>
<li>Use a higher level.</li>
</ol>
<ol start="10" style="list-style-type: decimal">
<li>Set the seed at 1, then use the <code>replicate</code> function to
repeat the code used in exercise 9 10,000 times. What proportion of the
time do we reject at the 0.05 level?</li>
<li>Note that, not surprisingly, the power is lower than 10%. Repeat the
exercise above for samples sizes of 30, 60, 90 and 120. Which of those
four gives you power of about 80%?</li>
<li>Repeat problem 11, but now require an <i>α</i> level of 0.01. Which
of those four gives you power of about 80%?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Solution to Exercises
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li><p><code>set.seed(1)</code><br><code>dat.ns &lt;- sample(bwt.nonsmoke, size = 25)</code><br><code>dat.s &lt;- sample(bwt.smoke, size = 25)</code><br><code>tval &lt;- t.test(bwt.smoke, bwt.nonsmoke)$p.value</code></p></li>
<li>
</li>
<li><p><code>se &lt;- sd(dat.ns)/sqrt(N)</code><br><code>qnorm(1-0.01/2)*se</code></p></li>
<li><p><code>qt(1-0.01/2, df=2*N-2)*se</code></p></li>
<li><p><code>set.seed(1)</code><br><code>dat.ns &lt;- sample(bwt.nonsmoke, size = 5)</code><br><code>dat.s &lt;- sample(bwt.smoke, size = 5)</code><br><code>tval &lt;- t.test(dat.ns, bdat.s)$p.value</code><br><code>tval</code></p></li>
</ol>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Power is the probability of detecting a real effect.</li>
<li>Effect size with a confidence interval is a more meaningful measure
than a p-value.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-inference-montecarlo"><p>Content from <a href="inference-montecarlo.html">Monte Carlo simulation</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-montecarlo.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How are Monte Carlo simulations used in practice?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Generate simulated data to test out hypotheses.</li>
<li>Plot distributions and determine whether they approximate the CLT or
t-distribution.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="monte-carlo-simulation">Monte Carlo Simulation<a class="anchor" aria-label="anchor" href="#monte-carlo-simulation"></a>
</h2>
<hr class="half-width">
<p>Computers can be used to generate pseudo-random numbers. For
practical purposes these pseudo-random numbers can be used to imitate
random variables from the real world. This permits us to examine
properties of random variables using a computer instead of theoretical
or analytical derivations. One very useful aspect of this concept is
that we can create <em>simulated</em> data to test out ideas or
competing methods, without actually having to perform laboratory
experiments.</p>
<p>Simulations can also be used to check theoretical or analytical
results. Also, many of the theoretical results we use in statistics are
based on asymptotics: they hold when the sample size goes to infinity.
In practice, we never have an infinite number of samples so we may want
to know how well the theory works with our actual sample size. Sometimes
we can answer this question analytically, but not always. Simulations
are extremely useful in these cases.</p>
<p>As an example, let’s use a Monte Carlo simulation to compare the CLT
to the t-distribution approximation for different sample sizes.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"./data/mice_pheno.csv"</span><span class="op">)</span></span>
<span><span class="va">controlPopulation</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pheno</span>, <span class="va">Sex</span> <span class="op">==</span> <span class="st">"F"</span> <span class="op">&amp;</span> <span class="va">Diet</span> <span class="op">==</span> <span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span>  </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span></code></pre>
</div>
<p>We will build a function that automatically generates a t-statistic
under the null hypothesis for a sample size of <code>n</code>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ttestgenerator</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># note that here we have a false "high fat" group where we actually</span></span>
<span>  <span class="co"># sample from the chow or control population. </span></span>
<span>  <span class="co"># This is because we are modeling the null.</span></span>
<span>  <span class="va">cases</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">controlPopulation</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="va">tstat</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">cases</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">controls</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> </span>
<span>      <span class="fu">sqrt</span><span class="op">(</span> <span class="fu">var</span><span class="op">(</span><span class="va">cases</span><span class="op">)</span><span class="op">/</span><span class="va">n</span> <span class="op">+</span> <span class="fu">var</span><span class="op">(</span><span class="va">controls</span><span class="op">)</span><span class="op">/</span><span class="va">n</span> <span class="op">)</span> </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">tstat</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="va">ttests</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">1000</span>, <span class="fu">ttestgenerator</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>With 1,000 Monte Carlo simulated occurrences of this random variable,
we can now get a glimpse of its distribution:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">ttests</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-montecarlo-rendered-ttest_hist-1.png" alt="Histogram of 1000 Monte Carlo simulated t-statistics." class="figure mx-auto d-block"><figcaption>
Histogram of 1000 Monte Carlo simulated t-statistics.
</figcaption></figure><p>So is the distribution of this t-statistic well approximated by the
normal distribution? In the next chapter, we will formally introduce
quantile-quantile plots, which provide a useful visual inspection of how
well one distribution approximates another. As we will explain later, if
points fall on the identity line, it means the approximation is a good
one.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">ttests</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-montecarlo-rendered-ttest_qqplot-1.png" alt="Quantile-quantile plot comparing 1000 Monte Carlo simulated t-statistics to theoretical normal distribution." class="figure mx-auto d-block"><figcaption>
Quantile-quantile plot comparing 1000 Monte Carlo simulated t-statistics
to theoretical normal distribution.
</figcaption></figure><p>This looks like a very good approximation. For this particular
population, a sample size of 10 was large enough to use the CLT
approximation. How about 3?</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ttests</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">1000</span>, <span class="fu">ttestgenerator</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">ttests</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-montecarlo-rendered-unnamed-chunk-6-1.png" alt="Quantile-quantile plot comparing 1000 Monte Carlo simulated t-statistics with three degrees of freedom to theoretical normal distribution." class="figure mx-auto d-block"><figcaption>
Quantile-quantile plot comparing 1000 Monte Carlo simulated t-statistics
with three degrees of freedom to theoretical normal distribution.
</figcaption></figure><p>Now we see that the large quantiles, referred to by statisticians as
the <em>tails</em>, are larger than expected (below the line on the left
side of the plot and above the line on the right side of the plot). In
the previous module, we explained that when the sample size is not large
enough and the <em>population values</em> follow a normal distribution,
then the t-distribution is a better approximation. Our simulation
results seem to confirm this:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">999</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">/</span><span class="fl">1000</span></span>
<span><span class="fu">qqplot</span><span class="op">(</span><span class="fu">qt</span><span class="op">(</span><span class="va">ps</span>, df <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="fl">3</span> <span class="op">-</span> <span class="fl">2</span><span class="op">)</span>, <span class="va">ttests</span>, xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>,<span class="fl">6</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>,<span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-montecarlo-rendered-unnamed-chunk-7-1.png" alt="Quantile-quantile plot comparing 1000 Monte Carlo simulated t-statistics with three degrees of freedom to theoretical t-distribution." class="figure mx-auto d-block"><figcaption>
Quantile-quantile plot comparing 1000 Monte Carlo simulated t-statistics
with three degrees of freedom to theoretical t-distribution.
</figcaption></figure><p>The t-distribution is a much better approximation in this case, but
it is still not perfect. This is due to the fact that the original data
is not that well approximated by the normal distribution.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">controlPopulation</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-montecarlo-rendered-unnamed-chunk-8-1.png" alt="Quantile-quantile of original data compared to theoretical quantile distribution." class="figure mx-auto d-block"><figcaption>
Quantile-quantile of original data compared to theoretical quantile
distribution.
</figcaption></figure></section><section><h2 class="section-heading" id="parametric-simulations-for-the-observations">Parametric Simulations for the Observations<a class="anchor" aria-label="anchor" href="#parametric-simulations-for-the-observations"></a>
</h2>
<hr class="half-width">
<p>The technique we used to motivate random variables and the null
distribution was a type of Monte Carlo simulation. We had access to
population data and generated samples at random. In practice, we do not
have access to the entire population. The reason for using the approach
here was for educational purposes. However, when we want to use Monte
Carlo simulations in practice, it is much more typical to assume a
parametric distribution and generate a population from this, which is
called a <em>parametric simulation</em>. This means that we take
parameters estimated from the real data (here the mean and the standard
deviation), and plug these into a model (here the normal distribution).
This is actually the most common form of Monte Carlo simulation.</p>
<p>For the case of weights, we could use our knowledge that mice
typically weigh 24 grams with a SD of about 3.5 grams, and that the
distribution is approximately normal, to generate population data:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">controls</span><span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">5000</span>, mean<span class="op">=</span><span class="fl">24</span>, sd<span class="op">=</span><span class="fl">3.5</span><span class="op">)</span> </span></code></pre>
</div>
<p>After we generate the data, we can then repeat the exercise above. We
no longer have to use the <code>sample</code> function since we can
re-generate random normal numbers. The <code>ttestgenerator</code>
function therefore can be written as follows:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ttestgenerator</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span>, <span class="va">mean</span><span class="op">=</span><span class="fl">24</span>, <span class="va">sd</span><span class="op">=</span><span class="fl">3.5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cases</span> <span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span>, <span class="va">mean</span>, <span class="va">sd</span><span class="op">)</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span>, <span class="va">mean</span>, <span class="va">sd</span><span class="op">)</span></span>
<span>  <span class="va">tstat</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">cases</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">controls</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> </span>
<span>      <span class="fu">sqrt</span><span class="op">(</span> <span class="fu">var</span><span class="op">(</span><span class="va">cases</span><span class="op">)</span><span class="op">/</span><span class="va">n</span> <span class="op">+</span> <span class="fu">var</span><span class="op">(</span><span class="va">controls</span><span class="op">)</span><span class="op">/</span><span class="va">n</span> <span class="op">)</span> </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">tstat</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span></code></pre>
</div>
<blockquote>
<h2 id="exercises">Exercises</h2>
<p>We have used Monte Carlo simulation throughout this chapter to
demonstrate statistical concepts; namely, sampling from the population.
We mostly applied this to demonstrate the statistical properties related
to inference on differences in averages. Here, we will consider examples
of how Monte Carlo simulations are used in practice.</p>
<ol style="list-style-type: decimal">
<li>Imagine you are William Sealy Gosset and have just mathematically
derived the distribution of the t-statistic when the sample comes from a
normal distribution. Unlike Gosset you have access to computers and can
use them to check the results.<br>
Let’s start by creating an outcome. Set the seed at 1, use
<code>rnorm</code> to generate a random sample of size 5,
<i>X<sub>1</sub>…X<sub>5</sub></i> from a standard normal distribution,
then compute the t-statistic <i>t = √5</i>X̄<i>/s</i> with <i>/s</i> the
sample standard deviation. What value do you observe?</li>
<li>You have just performed a Monte Carlo simulation using
<code>rnorm</code>, a random number generator for normally distributed
data. Gosset’s mathematical calculation tells us that this random
variable follows a t-distribution with <i>N-1</i> degrees of freedom.
Monte Carlo simulations can be used to check the theory: we generate
many outcomes and compare them to the theoretical result.<br>
Set the seed to 1, generate B= 1000 t-statistics as done in exercise
1.<br>
What percent are larger than 2?</li>
<li>The answer to exercise 2 is very similar to the theoretical
prediction: 1-pt(2, df=4). We can check several such quantiles using the
<code>qqplot</code> function. To obtain quantiles for the t-distribution
we can generate percentiles from just above 0 to just below 1:B=100; ps
= seq(1/(B+1), 1-1/(B+1), len=B) and compute the quantiles with
<code>qt(ps, df=4)</code>. Now we can use <code>qqplot</code> to compare
these theoretical quantiles to those obtained in the Monte Carlo
simulation. Use Monte Carlo simulation developed for exercise 2 to
corroborate that the t-statistic t=pNX/s follows a t-distribution for
several values of N.<br>
For which sample sizes does the approximation best work?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Larger sample sizes.<br>
</li>
<li>Smaller sample sizes.<br>
</li>
<li>The approximations are spot on for all sample sizes.<br>
</li>
<li>None. We should use CLT instead.</li>
</ol>
<ol start="4" style="list-style-type: decimal">
<li>Use Monte Carlo simulation to corroborate that the t-statistic
comparing two means and obtained with normally distributed (mean 0 and
sd) data follows a t-distribution. In this case we will use the
<code>t.test</code> function with <code>var.equal=TRUE</code>. With this
argument the degrees of freedom will be df = 2 * N-2 with N the sample
size. For which sample sizes does the approximation best work?<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Larger sample sizes.<br>
</li>
<li>Smaller sample sizes.<br>
</li>
<li>The approximations are spot on for all sample sizes.<br>
</li>
<li>None. We should use CLT instead.</li>
</ol>
<ol start="5" style="list-style-type: decimal">
<li>Is the following statement true or false?<br>
If instead of generating the sample with <code>X=rnorm(15)</code>, we
generate it with binary data
<code>X = rbinom(n=15, size=1, prob=0.5)</code> then the t-statistic
<code>tstat &lt;- sqrt(15) * mean(X)/sd(X)</code> is approximated by a
t-distribution with 14 degrees of freedom.</li>
<li>Is the following statement true or false?<br>
If instead of generating the sample with <code>X = rnorm(N)</code> with
N=500, we generate the data with binary data
<code>X = rbinom(n=500, size=1, prob=0.5)</code>, then the t-statistic
<code>sqrt(N) * mean(X)/sd(X)</code> is approximated by a t-distribution
with 499 degrees of freedom.</li>
<li>We can derive approximation of the distribution of the sample
average or the t-statistic theoretically. However, suppose we are
interested in the distribution of a statistic for which a theoretical
approximation is not immediately obvious. Consider the sample median as
an example. Use a Monte Carlo to determine which of the following best
approximates the median of a sample taken from normally distributed
population with mean 0 and standard deviation 1.<br>
</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Just like for the average, the sample median is approximately normal
with mean 0 and SD1/pN.<br>
</li>
<li>The sample median is not approximately normal.<br>
</li>
<li>The sample median is t-distributed for small samples and normally
distributed for large ones.<br>
</li>
<li>The sample median is approximately normal with mean 0 and SD larger
than 1/pN.</li>
</ol>
</blockquote>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Random number generators can simulate imitate random variables from
the real world.</li>
<li>Simulated data lets us examine properties of random variables and
test out ideas or competing methods computationally.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-inference-permutations"><p>Content from <a href="inference-permutations.html">Permutations</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-permutations.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is a permutation test?</li>
<li>When is a permutation test helpful?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Generate a null distribution with permutations.</li>
<li>Evaluate the results of a permutation test.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="permutation-tests">Permutation Tests<a class="anchor" aria-label="anchor" href="#permutation-tests"></a>
</h2>
<hr class="half-width">
<p>Suppose we have a situation in which none of the standard
mathematical statistical approximations apply. We have computed a
summary statistic, such as the difference in mean, but do not have a
useful approximation, such as that provided by the CLT. In practice, we
do not have access to all values in the population so we can’t perform a
simulation as done above. Permutation tests can be useful in these
scenarios.</p>
<p>We are back to the scenario where we only have 10 measurements for
each group.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fWeights</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleMiceWeights.csv"</span><span class="op">)</span> <span class="co"># we read this data in earlier</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"chow"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">fWeights</span>, <span class="va">Diet</span><span class="op">==</span><span class="st">"hf"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">Bodyweight</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">unlist</span></span>
<span><span class="va">obsdiff</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<p>In previous sections, we showed parametric approaches that helped
determine if the observed difference was significant. Permutation tests
take advantage of the fact that if we randomly shuffle the cases and
control labels, then the null is true. So we shuffle the cases and
control labels and assume that the ensuing distribution approximates the
null distribution. Here is how we generate a null distribution by
shuffling the data 1,000 times:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">avgdiff</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">1000</span>, <span class="op">{</span></span>
<span>    <span class="va">all</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">control</span>, <span class="va">treatment</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">newcontrols</span> <span class="op">&lt;-</span> <span class="va">all</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span><span class="op">]</span></span>
<span>    <span class="va">newtreatments</span> <span class="op">&lt;-</span> <span class="va">all</span><span class="op">[</span><span class="op">(</span><span class="va">N</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">N</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">newtreatments</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">newcontrols</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">avgdiff</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="va">obsdiff</span>, col<span class="op">=</span><span class="st">"red"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-permutations-rendered-diff_hist-1.png" alt="Histogram of difference between averages from permutations. Vertical line shows the observed difference." class="figure mx-auto d-block"><figcaption>
Histogram of difference between averages from permutations. Vertical
line shows the observed difference.
</figcaption></figure><p>How many of the null means are bigger than the observed value? That
proportion would be the p-value for the null. We add a 1 to the
numerator and denominator to account for misestimation of the p-value
(for more details see <a href="https://www.ncbi.nlm.nih.gov/pubmed/21044043" class="external-link">Phipson and Smyth,
Permutation P-values should never be zero</a>).</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#the proportion of permutations with larger difference</span></span>
<span><span class="op">(</span><span class="fu">sum</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">avgdiff</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu">abs</span><span class="op">(</span><span class="va">obsdiff</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">avgdiff</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.05094905</code></pre>
</div>
<p>Now let’s repeat this experiment for a smaller dataset. We create a
smaller dataset by sampling:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">control</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="va">obsdiff</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">treatment</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<p>and repeat the exercise:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">avgdiff</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">1000</span>, <span class="op">{</span></span>
<span>    <span class="va">all</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">control</span>,<span class="va">treatment</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">newcontrols</span> <span class="op">&lt;-</span> <span class="va">all</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span><span class="op">]</span></span>
<span>    <span class="va">newtreatments</span> <span class="op">&lt;-</span> <span class="va">all</span><span class="op">[</span><span class="op">(</span><span class="va">N</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">N</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">newtreatments</span><span class="op">)</span> <span class="op">-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">newcontrols</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">avgdiff</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="va">obsdiff</span>, col<span class="op">=</span><span class="st">"red"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/inference-permutations-rendered-diff_hist_N50-1.png" alt="Histogram of difference between averages from permutations for smaller sample size. Vertical line shows the observed difference." class="figure mx-auto d-block"><figcaption>
Histogram of difference between averages from permutations for smaller
sample size. Vertical line shows the observed difference.
</figcaption></figure><p>Now the observed difference is not significant using this approach.
Keep in mind that there is no theoretical guarantee that the null
distribution estimated from permutations approximates the actual null
distribution. For example, if there is a real difference between the
populations, some of the permutations will be unbalanced and will
contain some samples that explain this difference. This implies that the
null distribution created with permutations will have larger tails than
the actual null distribution. This is why permutations result in
conservative p-values. For this reason, when we have few samples, we
can’t do permutations.</p>
<p>Note also that permutation tests still have assumptions: samples are
assumed to be independent and “exchangeable”. If there is hidden
structure in your data, then permutation tests can result in estimated
null distributions that underestimate the size of tails because the
permutations may destroy the existing structure in the original
data.</p>
<blockquote>
<p>Exercises We will use the following dataset to demonstrate the use of
permutations:<br><code>url&lt;-"https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/babies.txt"</code><br><code>filename &lt;- basename(url)</code><br><code>download(url, destfile=filename)</code><br><code>babies&lt;-read.table("babies.txt", header=TRUE)</code><br><code>bwt.nonsmoke &lt;- filter(babies, smoke==0) %&gt;% select(bwt) %&gt;% unlist</code><br><code>bwt.smoke &lt;- filter(babies, smoke==1) %&gt;% select(bwt) %&gt;% unlist</code></p>
<ol style="list-style-type: decimal">
<li>We will generate the following random variable based on a sample
size of 10 and observe the following difference:<br><code>N &lt;- 10</code><br><code>set.seed(1)</code><br><code>nonsmokers&lt;-sample(bwt.nonsmoke, N)</code><br><code>smokers&lt;-sample(bwt.smoke, N)</code><br><code>obs &lt;- mean(smokers) - mean(nonsmokers)</code><br>
The question is whether this observed difference is statistically
significant. We do not want to rely on the assumptions needed for the
normal or t-distribution approximations to hold, so instead we will use
permutations. We will reshuffle the data and recompute the mean. We can
create one permuted sample with the following code:<br><code>dat &lt;- c(smokers, nonsmokers)</code><br><code>shuffle &lt;- sample(dat)</code><br><code>smokersstar &lt;- shuffle[1:N]</code><br><code>nonsmokersstar &lt;- shuffle[(N+1):(2*N)]</code><br><code>mean(smokersstar) - mean(nonsmokersstar)</code><br>
The last value is one observation from the null distribution we will
construct. Set the seed at 1, and then repeat the permutation 1,000
times to create a null distribution. What is the permutation derived
p-value for our observation?</li>
<li>Repeat the above exercise, but instead of the differences in mean,
consider the differences in median.
<code>obs &lt;- median(smokers) - median(nonsmokers)</code> What is the
permutation based p-value?</li>
</ol>
</blockquote>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Permutation randomly shuffles data so that the null is true.</li>
<li>Permutation tests are useful when there is no good approximation
such as that provided by the CLT.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-inference-assoc-tests"><p>Content from <a href="inference-assoc-tests.html">Association tests</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/inference-assoc-tests.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the Chi-squared test?</li>
<li>What is Fisher’s exact test?</li>
<li>When would these tests be used?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe and execute the Chi-squared test.</li>
<li>Compute Fisher’s exact test.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="association-tests">Association Tests<a class="anchor" aria-label="anchor" href="#association-tests"></a>
</h2>
<hr class="half-width">
<p>The statistical tests we have covered up to now leave out a
substantial portion of life science projects. Specifically, we are
referring to data that is binary, categorical and ordinal. To give a
very specific example, consider genetic data where you have two groups
of genotypes (AA/Aa or aa) for cases and controls for a given disease.
The statistical question is if genotype and disease are associated. As
in the examples we have been studying previously, we have two
populations (AA/Aa and aa) and then numeric data for each, where disease
status can be coded as 0 or 1. So why can’t we perform a t-test? Note
that the data is either 0 (control) or 1 (cases). It is pretty clear
that this data is not normally distributed so the t-distribution
approximation is certainly out of the question. We could use CLT if the
sample size is large enough; otherwise, we can use <em>association
tests</em>.</p>
<div class="section level4">
<h4 id="lady-tasting-tea">Lady Tasting Tea<a class="anchor" aria-label="anchor" href="#lady-tasting-tea"></a>
</h4>
<p>One of the most famous examples of hypothesis testing was performed
by <a href="https://en.wikipedia.org/wiki/Ronald_Fisher" class="external-link">R.A.
Fisher</a>. An acquaintance of Fisher’s claimed that she could tell if
milk was added before or after tea was poured. Fisher gave her four
pairs of cups of tea: one with milk poured first, the other after. The
order was randomized. Say she picked 3 out of 4 correctly, do we believe
she has a special ability? Hypothesis testing helps answer this question
by quantifying what happens by chance. This example is called the “Lady
Tasting Tea” experiment (and, as it turns out, Fisher’s friend was a
scientist herself, <a href="https://en.wikipedia.org/wiki/Muriel_Bristol" class="external-link">Muriel
Bristol</a>).</p>
<p>The basic question we ask is: if the tester is actually guessing,
what are the chances that she gets 3 or more correct? Just as we have
done before, we can compute a probability under the null hypothesis that
she is guessing 4 of each. If we assume this null hypothesis, we can
think of this particular example as picking 4 balls out of an urn with 4
green (correct answer) and 4 red (incorrect answer) balls.</p>
<p>Under the null hypothesis that she is simply guessing, each ball has
the same chance of being picked. We can then use combinatorics to figure
out each probability. The probability of picking 3 is <span class="math inline">\({4 \choose 3} {4 \choose 1} / {8 \choose 4} =
16/70\)</span>. The probability of picking all 4 correct is <span class="math inline">\({4 \choose 4} {4 \choose 0}/{8 \choose 4}=
1/70\)</span>. Thus, the chance of observing a 3 or something more
extreme, under the null hypothesis, is <span class="math inline">\(\approx 0.24\)</span>. This is the p-value. The
procedure that produced this p-value is called <em>Fisher’s exact
test</em> and it uses the <em>hypergeometric distribution</em>.</p>
</div>
<div class="section level4">
<h4 id="two-by-two-tables">Two By Two Tables<a class="anchor" aria-label="anchor" href="#two-by-two-tables"></a>
</h4>
<p>The data from the experiment above can be summarized by a two by two
table:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Poured Before"</span>,<span class="st">"Poured After"</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Guessed before"</span>,<span class="st">"Guessed after"</span><span class="op">)</span></span>
<span><span class="va">tab</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>              Guessed before Guessed after
Poured Before              3             1
Poured After               1             3</code></pre>
</div>
<p>The function <code>fisher.test</code> performs the calculations above
and can be obtained like this:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fisher.test</span><span class="op">(</span><span class="va">tab</span>, alternative<span class="op">=</span><span class="st">"greater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Fisher's Exact Test for Count Data

data:  tab
p-value = 0.2429
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 0.3135693       Inf
sample estimates:
odds ratio
  6.408309 </code></pre>
</div>
</div>
<div class="section level4">
<h4 id="chi-square-test">Chi-square Test<a class="anchor" aria-label="anchor" href="#chi-square-test"></a>
</h4>
<p>Genome-wide association studies (GWAS) have become ubiquitous in
biology. One of the main statistical summaries used in these studies are
Manhattan plots. The y-axis of a Manhattan plot typically represents the
negative of log (base 10) of the p-values obtained for association tests
applied at millions of single nucleotide polymorphisms (SNP). The x-axis
is typically organized by chromosome (chromosome 1 to 22, X, Y, etc.).
These p-values are obtained in a similar way to the test performed on
the tea taster. However, in that example the number of green and red
balls is experimentally fixed and the number of answers given for each
category is also fixed. Another way to say this is that the sum of the
rows and the sum of the columns are fixed. This defines constraints on
the possible ways we can fill the two by two table and also permits us
to use the hypergeometric distribution. In general, this is not the
case. Nonetheless, there is another approach, the Chi-squared test,
which is described below.</p>
<p>Imagine we have 250 individuals, where some of them have a given
disease and the rest do not. We observe that 20% of the individuals that
are homozygous for the minor allele (aa) have the disease compared to
10% of the rest. Would we see this again if we picked another 250
individuals?</p>
<p>Let’s create a dataset with these percentages:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">disease</span><span class="op">=</span><span class="fu">factor</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">180</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">20</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">40</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span>               labels<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"cases"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">genotype</span><span class="op">=</span><span class="fu">factor</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">"AA/Aa"</span>,<span class="fl">200</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="st">"aa"</span>,<span class="fl">50</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                levels<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"AA/Aa"</span>,<span class="st">"aa"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="va">disease</span>, <span class="va">genotype</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="va">dat</span><span class="op">[</span><span class="fu">sample</span><span class="op">(</span><span class="fu">nrow</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span> <span class="co">#shuffle them up</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    disease genotype
249   cases       aa
68  control    AA/Aa
167 control    AA/Aa
129 control    AA/Aa
162 control    AA/Aa
215 control       aa</code></pre>
</div>
<p>To create the appropriate two by two table, we will use the function
<code>table</code>. This function tabulates the frequency of each level
in a factor. For example:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">genotype</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>genotype
AA/Aa    aa
  200    50 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">disease</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>disease
control   cases
    220      30 </code></pre>
</div>
<p>If you provide the function with two factors, it will tabulate all
possible pairs and thus create the two by two table:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu">table</span><span class="op">(</span><span class="va">genotype</span>,<span class="va">disease</span><span class="op">)</span></span>
<span><span class="va">tab</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        disease
genotype control cases
   AA/Aa     180    20
   aa         40    10</code></pre>
</div>
<p>Note that you can feed <code>table</code> <span class="math inline">\(n\)</span> factors and it will tabulate all <span class="math inline">\(n\)</span>-tables.</p>
<p>The typical statistics we use to summarize these results is the odds
ratio (OR). We compute the odds of having the disease if you are an
“aa”: 10/40, the odds of having the disease if you are an “AA/Aa”:
20/180, and take the ratio: <span class="math inline">\((10/40) /
(20/180)\)</span></p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">tab</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">]</span><span class="op">/</span><span class="va">tab</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">tab</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span><span class="op">/</span><span class="va">tab</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2.25</code></pre>
</div>
<p>To compute a p-value, we don’t use the OR directly. We instead assume
that there is no association between genotype and disease, and then
compute what we expect to see in each <em>cell</em> of the table (note:
this use of the word “cell” refers to elements in a matrix or table and
has nothing to do with biological cells). Under the null hypothesis, the
group with 200 individuals and the group with 50 individuals were each
randomly assigned the disease with the same probability. If this is the
case, then the probability of disease is:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">=</span><span class="fu">mean</span><span class="op">(</span><span class="va">disease</span> <span class="op">==</span> <span class="st">"cases"</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.12</code></pre>
</div>
<p>The expected table is therefore:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expected</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">p</span>,<span class="va">p</span><span class="op">)</span> <span class="op">*</span> <span class="fu">sum</span><span class="op">(</span><span class="va">genotype</span><span class="op">==</span><span class="st">"AA/Aa"</span><span class="op">)</span>,</span>
<span>                  <span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">p</span>,<span class="va">p</span><span class="op">)</span> <span class="op">*</span> <span class="fu">sum</span><span class="op">(</span><span class="va">genotype</span><span class="op">==</span><span class="st">"aa"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">dimnames</span><span class="op">(</span><span class="va">expected</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">dimnames</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span></span>
<span><span class="va">expected</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        disease
genotype control cases
   AA/Aa     176    24
   aa         44     6</code></pre>
</div>
<p>The Chi-square test uses an asymptotic result (similar to the CLT)
related to the sums of independent binary outcomes. Using this
approximation, we can compute the probability of seeing a deviation from
the expected table as big as the one we saw. The p-value for this table
is:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">chisq.test</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.08857435</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="large-samples-small-p-values">Large Samples, Small p-values<a class="anchor" aria-label="anchor" href="#large-samples-small-p-values"></a>
</h4>
<p>As mentioned earlier, reporting only p-values is not an appropriate
way to report the results of your experiment. Many genetic association
studies seem to overemphasize p-values. They have large sample sizes and
report impressively small p-values. Yet when one looks closely at the
results, we realize odds ratios are quite modest: barely bigger than 1.
In this case the difference of having genotype AA/Aa or aa might not
change an individual’s risk for a disease in an amount which is
<em>practically significant</em>, in that one might not change one’s
behavior based on the small increase in risk.</p>
<p>There is not a one-to-one relationship between the odds ratio and the
p-value. To demonstrate, we recalculate the p-value keeping all the
proportions identical, but increasing the sample size by 10, which
reduces the p-value substantially (as we saw with the t-test under the
alternative hypothesis):</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tab</span><span class="op">&lt;-</span><span class="va">tab</span><span class="op">*</span><span class="fl">10</span></span>
<span><span class="fu">chisq.test</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.219624e-09</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="confidence-intervals-for-the-odds-ratio">Confidence Intervals for the Odds Ratio<a class="anchor" aria-label="anchor" href="#confidence-intervals-for-the-odds-ratio"></a>
</h4>
<p>Computing confidence intervals for the OR is not mathematically
straightforward. Unlike other statistics, for which we can derive useful
approximations of their distributions, the OR is not only a ratio, but a
ratio of ratios. Therefore, there is no simple way of using, for
example, the CLT.</p>
<p>One approach is to use the theory of <em>generalized linear
models</em> which provides estimates of the log odds ratio, rather than
the OR itself, that can be shown to be asymptotically normal. Here we
provide R code without presenting the theoretical details (for further
details please see a reference on generalized linear models such as <a href="https://en.wikipedia.org/wiki/Generalized_linear_model" class="external-link">Wikipedia</a>
or <a href="https://books.google.com/books?hl=en&amp;lr=&amp;id=h9kFH2_FfBkC" class="external-link">McCullagh
and Nelder, 1989</a>):</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">glm</span><span class="op">(</span><span class="va">disease</span> <span class="op">~</span> <span class="va">genotype</span>, family<span class="op">=</span><span class="st">"binomial"</span>, data<span class="op">=</span><span class="va">dat</span><span class="op">)</span></span>
<span><span class="va">coeftab</span> <span class="op">&lt;-</span> <span class="fu">summary</span><span class="op">(</span><span class="va">fit</span><span class="op">)</span><span class="op">$</span><span class="va">coef</span></span>
<span><span class="va">coeftab</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>              Estimate Std. Error   z value     Pr(&gt;|z|)
(Intercept) -2.1972246  0.2356828 -9.322803 1.133070e-20
genotypeaa   0.8109302  0.4249074  1.908487 5.632834e-02</code></pre>
</div>
<p>The second row of the table shown above gives you the estimate and SE
of the log odds ratio. Mathematical theory tells us that this estimate
is approximately normally distributed. We can therefore form a
confidence interval and then exponentiate to provide a confidence
interval for the OR.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ci</span> <span class="op">&lt;-</span> <span class="va">coeftab</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="va">coeftab</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu">exp</span><span class="op">(</span><span class="va">ci</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.9618616 5.2632310</code></pre>
</div>
<p>The confidence includes 1, which is consistent with the p-value being
bigger than 0.05. Note that the p-value shown here is based on a
different approximation to the one used by the Chi-square test, which is
why they differ.</p>
<p>Exercises We showed how to calculate a Chi-square test from a table.
Here we will show how to generate the table from data which is in the
form of a dataframe, so that you can then perform an association test to
see if two columns have an enrichment (or depletion) of shared
occurrences. Download the <a href="https://studio.edx.org/c4x/HarvardX/PH525.1x/asset/assoctest.csv" class="external-link">file</a>
into your working directory, and then read it into R:
<code>d = read.csv("assoctest.csv")</code>.<br>
1. This dataframe reflects the allele status (either AA/Aa or aa) and
the case/control status for 72 individuals. Compute the Chi-square test
for the association of genotype with case/control status (using the
<code>table</code> function and the <code>chisq.test</code> function).
Examine the table to see if there appears to be an association. What is
the X-squared statistic?<br>
2.Compute Fisher’s exact test <code>fisher.test</code> for the same
table. What is the p-value?</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Binary, categorical and ordinal data require different tests.</li>
<li>Chi-squared and Fisher’s exact tests are useful in these cases.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-eda-intro"><p>Content from <a href="eda-intro.html">Exploratory Data Analysis</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/eda-intro.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can data be visualized to reveal important relationships?</li>
<li>What is exploratory data analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Use a quantile-quantile plot to evaluate normality.</li>
<li>Evaluate data distribution with a boxplot.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="exploratory-data-analysis">Exploratory Data Analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h2>
<hr class="half-width">
<p>“The greatest value of a picture is when it forces us to notice what
we never expected to see.” -John W. Tukey</p>
<p>Biases, systematic errors and unexpected variability are common in
data from the life sciences. Failure to discover these problems often
leads to flawed analyses and false discoveries. As an example, consider
that experiments sometimes fail and not all data processing pipelines,
such as the <code>t.test</code> function in R, are designed to detect
these. Yet, these pipelines still give you an answer. Furthermore, it
may be hard or impossible to notice an error was made just from the
reported results.</p>
<p>Graphing data is a powerful approach to detecting these problems. We
refer to this as <em>exploratory data analysis</em> (EDA). Many
important methodological contributions to existing techniques in data
analysis were initiated by discoveries made via EDA. In addition, EDA
can lead to interesting biological discoveries which would otherwise be
missed through simply subjecting the data to a battery of hypothesis
tests. Through this book, we make use of exploratory plots to motivate
the analyses we choose. Here we present a general introduction to EDA
using height data.</p>
<p>We have already introduced some EDA approaches for
<em>univariate</em> data, namely the histograms and qq-plot. Here we
describe the qq-plot in more detail and some EDA and summary statistics
for paired data. We also give a demonstration of commonly used figures
that we recommend against.</p>
</section><section><h2 class="section-heading" id="quantile-quantile-plots">Quantile Quantile Plots<a class="anchor" aria-label="anchor" href="#quantile-quantile-plots"></a>
</h2>
<hr class="half-width">
<p>To corroborate that a theoretical distribution, for example the
normal distribution, is in fact a good approximation, we can use
quantile-quantile plots (qq-plots). Quantiles are best understood by
considering the special case of percentiles. The p-th percentile of a
list of a distribution is defined as the number q that is bigger than p%
of numbers (so the inverse of the cumulative distribution function we
defined earlier). For example, the median 50-th percentile is the
median. We can compute the percentiles for our list of heights:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="va">father.son</span> <span class="op">&lt;-</span> <span class="fu">UsingR</span><span class="fu">::</span><span class="va">father.son</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">father.son</span><span class="op">$</span><span class="va">fheight</span></span></code></pre>
</div>
<p>and for the normal distribution:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="op">(</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">99</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span> <span class="op">)</span><span class="op">/</span><span class="fl">100</span> </span>
<span><span class="va">qs</span> <span class="op">&lt;-</span> <span class="fu">quantile</span><span class="op">(</span><span class="va">x</span>, <span class="va">ps</span><span class="op">)</span></span>
<span><span class="va">normalqs</span> <span class="op">&lt;-</span> <span class="fu">qnorm</span><span class="op">(</span><span class="va">ps</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fu">popsd</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">normalqs</span>, <span class="va">qs</span>, xlab<span class="op">=</span><span class="st">"Normal percentiles"</span>, ylab<span class="op">=</span><span class="st">"Height percentiles"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span> <span class="co">##identity line</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-qqplot_example1-1.png" alt="First example of qqplot. Here we compute the theoretical quantiles ourselves." class="figure mx-auto d-block"><figcaption>
First example of qqplot. Here we compute the theoretical quantiles
ourselves.
</figcaption></figure><p>Note how close these values are. Also, note that we can see these
qq-plots with less code (this plot has more points than the one we
constructed manually, and so tail-behavior can be seen more
clearly).</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> </span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-qqplot_example2-1.png" alt="Second example of qqplot. Here we use the function qqnorm which computes the theoretical normal quantiles automatically." class="figure mx-auto d-block"><figcaption>
Second example of qqplot. Here we use the function qqnorm which computes
the theoretical normal quantiles automatically.
</figcaption></figure><p>However, the <code>qqnorm</code> function plots against a standard
normal distribution. This is why the line has slope
<code>popsd(x)</code> and intercept <code>mean(x)</code>.</p>
<p>In the example above, the points match the line very well. In fact,
we can run Monte Carlo simulations to see plots like this for data known
to be normally distributed.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span><span class="fl">1000</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-qqnorm_example-1.png" alt="Example of the qqnorm function. Here we apply it to numbers generated to follow a normal distribution." class="figure mx-auto d-block"><figcaption>
Example of the qqnorm function. Here we apply it to numbers generated to
follow a normal distribution.
</figcaption></figure><p>We can also get a sense for how non-normally distributed data will
look in a qq-plot. Here we generate data from the t-distribution with
different degrees of freedom. Notice that the smaller the degrees of
freedom, the fatter the tails. We call these “fat tails” because if we
plotted an empirical density or histogram, the density at the extremes
would be higher than the theoretical curve. In the qq-plot, this can be
seen in that the curve is lower than the identity line on the left side
and higher on the right side. This means that there are more extreme
values than predicted by the theoretical density plotted on the
x-axis.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dfs</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">12</span>, <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">df</span> <span class="kw">in</span> <span class="va">dfs</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">rt</span><span class="op">(</span><span class="fl">1000</span>, <span class="va">df</span><span class="op">)</span></span>
<span>  <span class="fu">qqnorm</span><span class="op">(</span><span class="va">x</span>, xlab<span class="op">=</span><span class="st">"t quantiles"</span>, main<span class="op">=</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"d.f="</span>,<span class="va">df</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>,<span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">qqline</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-qqnorm_of_t-1.png" alt="We generate t-distributed data for four degrees of freedom and make qqplots against normal theoretical quantiles." class="figure mx-auto d-block"><figcaption>
We generate t-distributed data for four degrees of freedom and make
qqplots against normal theoretical quantiles.
</figcaption></figure></section><section><h2 class="section-heading" id="boxplots">Boxplots<a class="anchor" aria-label="anchor" href="#boxplots"></a>
</h2>
<hr class="half-width">
<p>Data is not always normally distributed. Income is a widely cited
example. In these cases, the average and standard deviation are not
necessarily informative since one can’t infer the distribution from just
these two numbers. The properties described above are specific to the
normal. For example, the normal distribution does not seem to be a good
approximation for the direct compensation for 199 United States CEOs in
the year 2000.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exec.pay</span> <span class="op">&lt;-</span> <span class="fu">UsingR</span><span class="fu">::</span><span class="va">exec.pay</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">exec.pay</span><span class="op">)</span> </span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">exec.pay</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">exec.pay</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-execpay-1.png" alt="Histogram and QQ-plot of executive pay." class="figure mx-auto d-block"><figcaption>
Histogram and QQ-plot of executive pay.
</figcaption></figure><p>In addition to qq-plots, a practical summary of data is to compute 3
percentiles: 25th, 50th (the median) and the 75th. A boxplot shows these
3 values along with a range of the points within median ± 1.5 (75th
percentile - 25th percentile). Values outside this range are shown as
points and sometimes referred to as <em>outliers</em>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">boxplot</span><span class="op">(</span><span class="va">exec.pay</span>, ylab<span class="op">=</span><span class="st">"10,000s of dollars"</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">400</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-unnamed-chunk-2-1.png" alt="Simple boxplot of executive pay." class="figure mx-auto d-block"><figcaption>
Simple boxplot of executive pay.
</figcaption></figure><p>Here we show just one boxplot. However, one of the great benefits of
boxplots is that we could easily show many distributions in one plot, by
lining them up, side by side. We will see several examples of this
throughout the book.</p>
</section><section><h2 class="section-heading" id="scatterplots-and-correlation">Scatterplots and Correlation<a class="anchor" aria-label="anchor" href="#scatterplots-and-correlation"></a>
</h2>
<hr class="half-width">
<p>The methods described above relate to <em>univariate</em> variables.
In the biomedical sciences, it is common to be interested in the
relationship between two or more variables. A classic example is the
father/son height data used by <a href="https://en.wikipedia.org/wiki/Francis_Galton" class="external-link">Francis Galton</a>
to understand heredity. If we were to summarize these data, we could use
the two averages and two standard deviations since both distributions
are well approximated by the normal distribution. This summary, however,
fails to describe an important characteristic of the data.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">father.son</span> <span class="op">&lt;-</span> <span class="fu">UsingR</span><span class="fu">::</span><span class="va">father.son</span></span>
<span><span class="va">x</span><span class="op">=</span><span class="va">father.son</span><span class="op">$</span><span class="va">fheight</span></span>
<span><span class="va">y</span><span class="op">=</span><span class="va">father.son</span><span class="op">$</span><span class="va">sheight</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span>, xlab<span class="op">=</span><span class="st">"Father's height in inches"</span>, </span>
<span>     ylab<span class="op">=</span><span class="st">"Son's height in inches"</span>, </span>
<span>     main<span class="op">=</span><span class="fu">paste</span><span class="op">(</span><span class="st">"correlation ="</span>,<span class="fu">signif</span><span class="op">(</span><span class="fu">cor</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-scatterplot-1.png" alt="Heights of father and son pairs plotted against each other." class="figure mx-auto d-block"><figcaption>
Heights of father and son pairs plotted against each other.
</figcaption></figure><p>The scatter plot shows a general trend: the taller the father, the
taller the son. A summary of this trend is the correlation coefficient,
which in this case is 0.5. We will motivate this statistic by trying to
predict the son’s height using the father’s height.</p>
</section><section><h2 class="section-heading" id="stratification">Stratification<a class="anchor" aria-label="anchor" href="#stratification"></a>
</h2>
<hr class="half-width">
<p>Suppose we are asked to guess the height of randomly selected sons.
The average height, 68.7 inches, is the value with the highest
proportion (see histogram) and would be our prediction. But what if we
are told that the father is 72 inches tall, do we still guess 68.7?</p>
<p>The father is taller than average. Specifically, he is 1.75 standard
deviations taller than the average father. So should we predict that the
son is also 1.75 standard deviations taller? It turns out that this
would be an overestimate. To see this, we look at all the sons with
fathers who are about 72 inches. We do this by <em>stratifying</em> the
father heights.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">y</span>,<span class="fu">round</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">groups</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-boxplot-1.png" alt="Boxplot of son heights stratified by father heights." class="figure mx-auto d-block"><figcaption>
Boxplot of son heights stratified by father heights.
</figcaption></figure><div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">print</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span> <span class="fu">round</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">==</span> <span class="fl">72</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 70.67719</code></pre>
</div>
<p>Stratification followed by boxplots lets us see the distribution of
each group. The average height of sons with fathers that are 72 inches
tall is 70.7 inches. We also see that the <em>medians</em> of the strata
appear to follow a straight line (remember the middle line in the
boxplot shows the median, not the mean). This line is similar to the
<em>regression line</em>, with a slope that is related to the
correlation, as we will learn below.</p>
</section><section><h2 class="section-heading" id="bivariate-normal-distribution">Bivariate Normal Distribution<a class="anchor" aria-label="anchor" href="#bivariate-normal-distribution"></a>
</h2>
<hr class="half-width">
<p>Correlation is a widely used summary statistic in the life sciences.
However, it is often misused or misinterpreted. To properly interpret
correlation we actually have to understand the bivariate normal
distribution.</p>
<p>A pair of random variables <i>(X,Y)</i> is considered to be
approximated by bivariate normal when the proportion of values below,
for example <i>a</i> and <i>b</i>, is approximated by this
expression:</p>
<figure><img src="fig/12-probXltaYltb.png" class="figure mx-auto d-block"></figure><figure><img src="fig/12-integralsMinusInfab.png" class="figure mx-auto d-block"></figure><p>This may seem like a rather complicated equation, but the concept
behind it is rather intuitive. An alternative definition is the
following: fix a value <i>x</i> and look at all the pairs <i>(X,Y)</i>
for which <i>X=x</i>. Generally, in statistics we call this exercise
<em>conditioning</em>. We are conditioning <i>Y</i> on <i>X</i>. If a
pair of random variables is approximated by a bivariate normal
distribution, then the distribution of <i>Y</i> conditioned on
<i>X=x</i> is approximated with a normal distribution, no matter what
<i>x</i> we choose. Let’s see if this holds with our height data. We
show 4 different strata:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu">split</span><span class="op">(</span><span class="va">y</span>,<span class="fu">round</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">c</span><span class="op">(</span><span class="fl">5</span>,<span class="fl">8</span>,<span class="fl">11</span>,<span class="fl">14</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">qqnorm</span><span class="op">(</span><span class="va">groups</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,main<span class="op">=</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"X="</span>,<span class="fu">names</span><span class="op">(</span><span class="va">groups</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,<span class="st">" strata"</span><span class="op">)</span>,</span>
<span>         ylim<span class="op">=</span><span class="fu">range</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">2.5</span>,<span class="fl">2.5</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">qqline</span><span class="op">(</span><span class="va">groups</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-qqnorm_of_strata-1.png" alt="qqplots of son heights for four strata defined by father heights." class="figure mx-auto d-block"><figcaption>
qqplots of son heights for four strata defined by father heights.
</figcaption></figure><p>Now we come back to defining correlation. Mathematical statistics
tells us that when two variables follow a bivariate normal distribution,
then for any given value of <i>x</i>, the average of the <i>Y</i> in
pairs for which <i>X=x</i> is:</p>
<figure><img src="fig/12-muYplusRho.png" class="figure mx-auto d-block"></figure><p>Note that this is a line with slope</p>
<figure><img src="fig/12-rhoSigmaYX.png" class="figure mx-auto d-block"></figure><p>This is referred to as the <em>regression line</em>. If the SDs are
the same, then the slope of the regression line is the correlation ρ.
Therefore, if we standardize <i>X</i> and <i>Y</i>, the correlation is
the slope of the regression line.</p>
<p>Another way to see this is to form a prediction Ŷ: for every SD away
from the mean in <i>x</i>, we predict ρ SDs away for <i>Y</i>:</p>
<figure><img src="fig/12-YhatfracEqRhofrac.png" class="figure mx-auto d-block"></figure><p>If there is perfect correlation, we predict the same number of SDs.
If there is 0 correlation, then we don’t use <i>x</i> at all. For values
between 0 and 1, the prediction is somewhere in between. For negative
values, we simply predict in the opposite direction.</p>
<p>To confirm that the above approximations hold in this case, let’s
compare the mean of each strata to the identity line and the regression
line:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">=</span><span class="op">(</span> <span class="va">x</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">)</span><span class="op">/</span><span class="fu">sd</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="va">y</span><span class="op">=</span><span class="op">(</span> <span class="va">y</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">)</span><span class="op">/</span><span class="fu">sd</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">means</span><span class="op">=</span><span class="fu">tapply</span><span class="op">(</span><span class="va">y</span>, <span class="fu">round</span><span class="op">(</span><span class="va">x</span><span class="op">*</span><span class="fl">4</span><span class="op">)</span><span class="op">/</span><span class="fl">4</span>, <span class="va">mean</span><span class="op">)</span></span>
<span><span class="va">fatherheights</span><span class="op">=</span><span class="fu">as.numeric</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">means</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">fatherheights</span>, <span class="va">means</span>, ylab<span class="op">=</span><span class="st">"average of strata of son heights"</span>, ylim<span class="op">=</span><span class="fu">range</span><span class="op">(</span><span class="va">fatherheights</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>, <span class="fu">cor</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-intro-rendered-scatterplot2-1.png" alt="Average son height of each strata plotted against father heights defining the strata" class="figure mx-auto d-block"><figcaption>
Average son height of each strata plotted against father heights
defining the strata
</figcaption></figure><div class="section level4">
<h4 id="variance-explained">Variance explained<a class="anchor" aria-label="anchor" href="#variance-explained"></a>
</h4>
<p>The standard deviation of the <em>conditional</em> distribution
described above is:</p>
<figure><img src="fig/12-sqrt1minusRhoSq.png" class="figure mx-auto d-block"></figure><p>This is where statements like <i>X</i> explains ρ<sup>2</sup> × 100%
of the variation in <i>Y</i>: the variance of <i>Y</i> is σ<sup>2</sup>
and, once we condition, it goes down to (1-ρ<sup>2&lt;/sup)
σ<sup>2_Y</sup>. It is important to remember that the “variance
explained” statement only makes sense when the data is approximated by a
bivariate normal distribution.</sup></p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Given the above histogram, how many people are between the ages of 35
and 45?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Solution to Exercise 1
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>The <code>InsectSprays</code> data set is included in R. The dataset
reports the counts of insects in agricultural experimental units treated
with different insecticides. Make a boxplot and determine which
insecticide appears to be most effective.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Solution to Exercise 2
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Download and load this dataset into R. Use exploratory data analysis
tools to determine which two columns are different from the rest.</p>
<ol style="list-style-type: decimal">
<li>Which is the column that is positively skewed?</li>
<li>Which is the column that is negatively skewed?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Solution to Exercise 3
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Let’s consider a random sample of finishers from the New York City
Marathon in 2002. This dataset can be found in the UsingR package. Load
the library and then load the <code>nym.2002</code> dataset.
<code>library(UsingR)</code> <code><a href="https://dplyr.tidyverse.org" class="external-link">library(dplyr)</a></code>
<code>data(nym.2002, package="UsingR")</code></p>
<ol style="list-style-type: decimal">
<li>Use boxplots and histograms to compare the finishing times of males
and females. Which of the following best describes the difference?</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Males and females have the same distribution.</li>
<li>Most males are faster than most women.</li>
<li>Male and females have similar right skewed distributions with the
former, 20 minutes shifted to the left.</li>
<li>Both distribution are normally distributed with a difference in mean
of about 30 minutes.</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>Use <code>dplyr</code> to create two new data frames:
<code>males</code> and <code>females</code>, with the data for each
gender. For males, what is the Pearson correlation between age and time
to finish? For females, what is the Pearson correlation between age and
time to finish?</li>
<li>If we interpret these correlations without visualizing the data, we
would conclude that the older we get, the slower we run marathons,
regardless of gender. Look at scatterplots and boxplots of times
stratified by age groups (20-25, 25-30, etc..). After examining the
data, what is a more reasonable conclusion?</li>
</ol>
<ol style="list-style-type: upper-alpha">
<li>Finish times are constant up until about our 40s, then we get
slower.</li>
<li>On average, finish times go up by about 7 minutes every five
years.</li>
<li>The optimal age to run a marathon is 20-25.</li>
<li>Coding errors never happen: a five year old boy completed the 2012
NY city marathon.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Solution to Exercise 4
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Plot data to evaluate relationships between variables.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-eda-avoid-plots"><p>Content from <a href="eda-avoid-plots.html">Plots to avoid</a></p>
<hr>
<p>Last updated on 2025-08-15 |

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/episodes/eda-avoid-plots.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<hr>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="plots-to-avoid">Plots to Avoid<a class="anchor" aria-label="anchor" href="#plots-to-avoid"></a>
</h2>
<hr class="half-width">
<p>This section is based on a talk by <a href="https://kbroman.org/" class="external-link">Karl W. Broman</a> titled “How to Display
Data Badly,” in which he described how the default plots offered by
Microsoft Excel “obscure your data and annoy your readers” (<a href="https://kbroman.org/pages/talks.html" class="external-link">here</a> is a link to a
collection of Karl Broman’s talks). His lecture was inspired by the 1984
paper by H. Wainer: How to display data badly. American Statistician
38(2): 137–147. Dr. Wainer was the first to elucidate the principles of
the bad display of data. However, according to Karl Broman, “The now
widespread use of Microsoft Excel has resulted in remarkable advances in
the field.” Here we show examples of “bad plots” and how to improve them
in R.</p>
<div class="section level4">
<h4 id="general-principles">General principles<a class="anchor" aria-label="anchor" href="#general-principles"></a>
</h4>
<p>The aim of good data graphics is to display data accurately and
clearly. According to Karl Broman, some rules for displaying data
<em>badly</em> are:</p>
<ul>
<li>Display as little information as possible.</li>
<li>Obscure what you do show (with chart junk).</li>
<li>Use pseudo-3D and color gratuitously.</li>
<li>Make a pie chart (preferably in color and 3D).</li>
<li>Use a poorly chosen scale.</li>
<li>Ignore significant figures.</li>
</ul>
</div>
<div class="section level4">
<h4 id="pie-charts">Pie charts<a class="anchor" aria-label="anchor" href="#pie-charts"></a>
</h4>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">browsers</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span>Opera<span class="op">=</span><span class="fl">1</span>, Safari<span class="op">=</span><span class="fl">9</span>, Firefox<span class="op">=</span><span class="fl">20</span>, IE<span class="op">=</span><span class="fl">26</span>, Chrome<span class="op">=</span><span class="fl">44</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s say we want to report the results from a poll asking about
browser preference (taken in August 2013). The standard way of
displaying these is with a pie chart:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pie</span><span class="op">(</span><span class="va">browsers</span>, main<span class="op">=</span><span class="st">"Browser Usage (August 2013)"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-piechart-1.png" alt="Pie chart of browser usage." class="figure mx-auto d-block"><figcaption>
Pie chart of browser usage.
</figcaption></figure><p>Nonetheless, as stated by the help file for the <code>pie</code>
function:</p>
<p>“Pie charts are a very bad way of displaying information. The eye is
good at judging linear measures and bad at judging relative areas. A bar
chart or dot chart is a preferable way of displaying this type of
data.”</p>
<p>To see this, look at the figure above and try to determine the
percentages just from looking at the plot. Unless the percentages are
close to 25%, 50% or 75%, this is not so easy. Simply showing the
numbers is not only clear, but also saves on printing costs.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">browsers</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  Opera  Safari Firefox      IE  Chrome
      1       9      20      26      44 </code></pre>
</div>
<p>If you do want to plot them, then a barplot is appropriate. Here we
add horizontal lines at every multiple of 10 and then redraw the
bars:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="va">browsers</span>, </span>
<span>        main<span class="op">=</span><span class="st">"Browser Usage (August 2013)"</span>, </span>
<span>        ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">55</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span> <span class="op">*</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu">barplot</span><span class="op">(</span><span class="va">browsers</span>, add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-barplot-1.png" alt="Barplot of browser usage." class="figure mx-auto d-block"><figcaption>
Barplot of browser usage.
</figcaption></figure><p>Notice that we can now pretty easily determine the percentages by
following a horizontal line to the x-axis. Do avoid a 3D version since
it obfuscates the plot, making it more difficult to find the percentages
by eye.</p>
<p><img src="https://raw.githubusercontent.com/kbroman/Talk_Graphs/master/Figs/fig2b.png" alt="3D version." class="figure"> Even worse than pie charts are donut plots.</p>
<figure><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/1/11/Donut-Chart.svg/360px-Donut-Chart.svg.png" alt="Donut plot." class="figure mx-auto d-block"></figure><p>The reason is that by removing the center, we remove one of the
visual cues for determining the different areas: the angles. There is no
reason to ever use a donut plot to display data.</p>
</div>
<div class="section level4">
<h4 id="barplots-as-data-summaries">Barplots as data summaries<a class="anchor" aria-label="anchor" href="#barplots-as-data-summaries"></a>
</h4>
<p>While barplots are useful for showing percentages, they are
incorrectly used to display data from two groups being compared.
Specifically, barplots are created with height equal to the group means;
an antenna is added at the top to represent standard errors. This plot
is simply showing two numbers per group and the plot adds nothing:</p>
<figure><img src="https://raw.githubusercontent.com/kbroman/Talk_Graphs/master/Figs/fig1c.png" alt="Bad bar plots." class="figure mx-auto d-block"></figure><p>Much more informative is to summarize with a boxplot. If the number
of points is small enough, we might as well add them to the plot. When
the number of points is too large for us to see them, just showing a
boxplot is preferable. We can even set <code>range=0</code> in
<code>boxplot</code> to avoid drawing many outliers when the data is in
the range of millions.</p>
<p>Let’s recreate these barplots as boxplots. First let’s download the
data:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">downloader</span><span class="op">)</span></span>
<span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="st">"fig1.RData"</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/kbroman/Talk_Graphs/raw/master/R/fig1.RData"</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">file.exists</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span><span class="op">)</span> <span class="fu">download</span><span class="op">(</span><span class="va">url</span>, <span class="va">filename</span><span class="op">)</span></span>
<span><span class="fu">load</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can simply show the points and make simple boxplots:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>Treatment<span class="op">=</span><span class="va">x</span>, Control<span class="op">=</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">dat</span>,xlab<span class="op">=</span><span class="st">"Group"</span>, ylab<span class="op">=</span><span class="st">"Response"</span>, cex<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu">stripchart</span><span class="op">(</span><span class="va">dat</span>, vertical<span class="op">=</span><span class="cn">TRUE</span>, method<span class="op">=</span><span class="st">"jitter"</span>, pch<span class="op">=</span><span class="fl">16</span>, add<span class="op">=</span><span class="cn">TRUE</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-unnamed-chunk-4-1.png" alt="Treatment data and control data shown with a boxplot." class="figure mx-auto d-block"><figcaption>
Treatment data and control data shown with a boxplot.
</figcaption></figure><p>Notice how much more we see here: the center, spread, range, and the
points themselves. In the barplot, we only see the mean and the SE, and
the SE has more to do with sample size than with the spread of the
data.</p>
<p>This problem is magnified when our data has outliers or very large
tails. In the plot below, there appears to be very large and consistent
differences between the two groups:</p>
<figure><img src="https://raw.githubusercontent.com/kbroman/Talk_Graphs/master/Figs/fig3c.png" alt="Bar plots with outliers." class="figure mx-auto d-block"></figure><p>However, a quick look at the data demonstrates that this difference
is mostly driven by just two points. A version showing the data in the
log-scale is much more informative.</p>
<p>Start by downloading data:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">downloader</span><span class="op">)</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/kbroman/Talk_Graphs/raw/master/R/fig3.RData"</span></span>
<span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="st">"fig3.RData"</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">file.exists</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span><span class="op">)</span> <span class="fu">download</span><span class="op">(</span><span class="va">url</span>, <span class="va">filename</span><span class="op">)</span></span>
<span><span class="fu">load</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can show data and boxplots in original scale and
log-scale.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>Treatment<span class="op">=</span><span class="va">x</span>,Control<span class="op">=</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">dat</span>, xlab<span class="op">=</span><span class="st">"Group"</span>, ylab<span class="op">=</span><span class="st">"Response"</span>, cex<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu">stripchart</span><span class="op">(</span><span class="va">dat</span>, vertical<span class="op">=</span><span class="cn">TRUE</span>, method<span class="op">=</span><span class="st">"jitter"</span>, pch<span class="op">=</span><span class="fl">16</span>, add<span class="op">=</span><span class="cn">TRUE</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">dat</span>, xlab<span class="op">=</span><span class="st">"Group"</span>, ylab<span class="op">=</span><span class="st">"Response"</span>, log<span class="op">=</span><span class="st">"y"</span>, cex<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu">stripchart</span><span class="op">(</span><span class="va">dat</span>, vertical<span class="op">=</span><span class="cn">TRUE</span>, method<span class="op">=</span><span class="st">"jitter"</span>, pch<span class="op">=</span><span class="fl">16</span>, add<span class="op">=</span><span class="cn">TRUE</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-importance_of_log-1.png" alt="Data and boxplots for original data (left) and in log scale (right)." class="figure mx-auto d-block"><figcaption>
Data and boxplots for original data (left) and in log scale (right).
</figcaption></figure>
</div>
<div class="section level4">
<h4 id="show-the-scatter-plot">Show the scatter plot<a class="anchor" aria-label="anchor" href="#show-the-scatter-plot"></a>
</h4>
<p>The purpose of many statistical analyses is to determine
relationships between two variables. Sample correlations are typically
reported and sometimes plots are displayed to show this. However,
showing just the regression line is one way to display your data badly
since it hides the scatter. Surprisingly, plots such as the following
are commonly seen.</p>
<p>Again start by loading data:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/kbroman/Talk_Graphs/raw/master/R/fig4.RData"</span></span>
<span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="st">"fig4.RData"</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">file.exists</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span><span class="op">)</span> <span class="fu">download</span><span class="op">(</span><span class="va">url</span>, <span class="va">filename</span><span class="op">)</span></span>
<span><span class="fu">load</span><span class="op">(</span><span class="va">filename</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, lwd<span class="op">=</span><span class="fl">2</span>, type<span class="op">=</span><span class="st">"n"</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">y</span><span class="op">~</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span>,lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span>,<span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu">text</span><span class="op">(</span><span class="fl">78</span>, <span class="fl">200</span>, <span class="fu">paste</span><span class="op">(</span><span class="st">"y ="</span>, <span class="va">b</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"+"</span>, <span class="va">b</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"x"</span><span class="op">)</span>, adj<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rho</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fu">cor</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span> </span>
<span><span class="fu">text</span><span class="op">(</span><span class="fl">78</span>, <span class="fl">187</span>,<span class="fu">expression</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="va">rho</span>,<span class="st">" = 0.8567"</span><span class="op">)</span><span class="op">)</span>,adj<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span>,lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">y</span><span class="op">~</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span>,lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-show-data-1.png" alt="The plot on the left shows a regression line that was fitted to the data shown on the right. It is much more informative to show all the data." class="figure mx-auto d-block"><figcaption>
The plot on the left shows a regression line that was fitted to the data
shown on the right. It is much more informative to show all the data.
</figcaption></figure><p>When there are large amounts of points, the scatter can be shown by
binning in two dimensions and coloring the bins by the number of points
in the bin. An example of this is the <code>hexbin</code> function in
the <a href="https://cran.r-project.org/package=hexbin" class="external-link">hexbin
package</a>.</p>
</div>
<div class="section level4">
<h4 id="high-correlation-does-not-imply-replication">High correlation does not imply replication<a class="anchor" aria-label="anchor" href="#high-correlation-does-not-imply-replication"></a>
</h4>
<p>When new technologies or laboratory techniques are introduced, we are
often shown scatter plots and correlations from replicated samples. High
correlations are used to demonstrate that the new technique is
reproducible. Correlation, however, can be very misleading. Below is a
scatter plot showing data from replicated samples run on a high
throughput technology. This technology outputs 12,626 simultaneous
measurements.</p>
<p>In the plot on the left, we see the original data which shows very
high correlation. Yet the data follows a distribution with very fat
tails. Furthermore, 95% of the data is below the green line. The plot on
the right is in the log scale:</p>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-correlation-not-replication-1.png" alt="Gene expression data from two replicated samples. Left is in original scale and right is in log scale." class="figure mx-auto d-block"><figcaption>
Gene expression data from two replicated samples. Left is in original
scale and right is in log scale.
</figcaption></figure><p>Note that we do not show the code here as it is rather complex but we
explain how to make MA plots in a later chapter.</p>
<p>Although the correlation is reduced in the log-scale, it is very
close to 1 in both cases. Does this mean these data are reproduced? To
examine how well the second vector reproduces the first, we need to
study the differences. We therefore should plot that instead. In this
plot, we plot the difference (in the log scale) versus the average:</p>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-MAplot-1.png" alt="MA plot of the same data shown above shows that data is not replicated very well despite a high correlation." class="figure mx-auto d-block"><figcaption>
MA plot of the same data shown above shows that data is not replicated
very well despite a high correlation.
</figcaption></figure><p>These are referred to as Bland-Altman plots, or <em>MA plots</em> in
the genomics literature, and we will talk more about them later. “MA”
stands for “minus” and “average” because in this plot, the y-axis is the
difference between two samples on the log scale (the log ratio is the
difference of the logs), and the x-axis is the average of the samples on
the log scale. In this plot, we see that the typical difference in the
log (base 2) scale between two replicated measures is about 1. This
means that when measurements should be the same, we will, on average,
observe 2 fold difference. We can now compare this variability to the
differences we want to detect and decide if this technology is precise
enough for our purposes.</p>
</div>
<div class="section level4">
<h4 id="barplots-for-paired-data">Barplots for paired data<a class="anchor" aria-label="anchor" href="#barplots-for-paired-data"></a>
</h4>
<p>A common task in data analysis is the comparison of two groups. When
the dataset is small and data are paired, such as the outcomes before
and after a treatment, two-color barplots are unfortunately often used
to display the results.</p>
<figure><img src="https://raw.githubusercontent.com/kbroman/Talk_Graphs/master/Figs/fig6r_e.png" alt="Barplot for two variables." class="figure mx-auto d-block"></figure><p>There are better ways of showing these data to illustrate that there
is an increase after treatment. One is to simply make a scatter plot,
which shows that most points are above the identity line. Another
alternative is to plot the differences against the before values.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">12201970</span><span class="op">)</span></span>
<span><span class="va">before</span> <span class="op">&lt;-</span> <span class="fu">runif</span><span class="op">(</span><span class="fl">6</span>, <span class="fl">5</span>, <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">after</span> <span class="op">&lt;-</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">6</span>, <span class="va">before</span><span class="op">*</span><span class="fl">1.05</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">li</span> <span class="op">&lt;-</span> <span class="fu">range</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">before</span>, <span class="va">after</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ymx</span> <span class="op">&lt;-</span> <span class="fu">max</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">after</span><span class="op">-</span><span class="va">before</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">before</span>, <span class="va">after</span>, xlab<span class="op">=</span><span class="st">"Before"</span>, ylab<span class="op">=</span><span class="st">"After"</span>,</span>
<span>     ylim<span class="op">=</span><span class="va">li</span>, xlim<span class="op">=</span><span class="va">li</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>, lty<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">before</span>, <span class="va">after</span><span class="op">-</span><span class="va">before</span>, xlab<span class="op">=</span><span class="st">"Before"</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="va">ymx</span>, <span class="va">ymx</span><span class="op">)</span>,</span>
<span>     ylab<span class="op">=</span><span class="st">"Change (After - Before)"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">0</span>, lty<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-scatter-plot-for-two-vars-1.png" alt="For two variables a scatter plot or a 'rotated' plot similar to an MA plot is much more informative." class="figure mx-auto d-block"><figcaption>
For two variables a scatter plot or a ‘rotated’ plot similar to an MA
plot is much more informative.
</figcaption></figure><p>Line plots are not a bad choice, although I find them harder to
follow than the previous two. Boxplots show you the increase, but lose
the paired information.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">6</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">z</span>, <span class="fu">c</span><span class="op">(</span><span class="va">before</span>, <span class="va">after</span><span class="op">)</span>,</span>
<span>     xaxt<span class="op">=</span><span class="st">"n"</span>, ylab<span class="op">=</span><span class="st">"Response"</span>,</span>
<span>     xlab<span class="op">=</span><span class="st">""</span>, xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">0.5</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">axis</span><span class="op">(</span>side<span class="op">=</span><span class="fl">1</span>, at<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"Before"</span>,<span class="st">"After"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">segments</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">6</span><span class="op">)</span>, <span class="va">before</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">6</span><span class="op">)</span>, <span class="va">after</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span>     </span>
<span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">before</span>,<span class="va">after</span>,names<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"Before"</span>,<span class="st">"After"</span><span class="op">)</span>,ylab<span class="op">=</span><span class="st">"Response"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-lines-plot-box-plot-1.png" alt="Another alternative is a line plot. If we don't care about pairings, then the boxplot is appropriate." class="figure mx-auto d-block"><figcaption>
Another alternative is a line plot. If we don’t care about pairings,
then the boxplot is appropriate.
</figcaption></figure>
</div>
<div class="section level4">
<h4 id="gratuitous-3d">Gratuitous 3D<a class="anchor" aria-label="anchor" href="#gratuitous-3d"></a>
</h4>
<p>The figure below shows three curves. Pseudo 3D is used, but it is not
clear why. Maybe to separate the three curves? Notice how difficult it
is to determine the values of the curves at any given point:</p>
<figure><img src="https://raw.githubusercontent.com/kbroman/Talk_Graphs/master/Figs/fig8b.png" alt="Gratuitous 3-D." class="figure mx-auto d-block"></figure><p>This plot can be made better by simply using color to distinguish the
three lines:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##First read data</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/kbroman/Talk_Graphs/raw/master/R/fig8dat.csv"</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="va">url</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##Now make alternative plot</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,xlab<span class="op">=</span><span class="st">"log Dose"</span>,ylab<span class="op">=</span><span class="st">"Proportion survived"</span>,ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>     type<span class="op">=</span><span class="st">"l"</span>,lwd<span class="op">=</span><span class="fl">2</span>,col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="va">x</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>,lwd<span class="op">=</span><span class="fl">2</span>,col<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="va">x</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>,lwd<span class="op">=</span><span class="fl">2</span>,col<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0.4</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"Drug A"</span>,<span class="st">"Drug B"</span>,<span class="st">"Drug C"</span><span class="op">)</span>,lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-colors-for-different-lines-1.png" alt="This plot demonstrates that using color is more than enough to distinguish the three lines." class="figure mx-auto d-block"><figcaption>
This plot demonstrates that using color is more than enough to
distinguish the three lines.
</figcaption></figure>
</div>
<div class="section level4">
<h4 id="ignoring-important-factors">Ignoring important factors<a class="anchor" aria-label="anchor" href="#ignoring-important-factors"></a>
</h4>
<p>In this example, we generate data with a simulation. We are studying
a dose-response relationship between two groups: treatment and control.
We have three groups of measurements for both control and treatment.
Comparing treatment and control using the common barplot.</p>
<figure><img src="https://raw.githubusercontent.com/kbroman/Talk_Graphs/master/Figs/fig9d.png" alt="Ignoring important factors." class="figure mx-auto d-block"></figure><p>Instead, we should show each curve. We can use color to distinguish
treatment and control, and dashed and solid lines to distinguish the
original data from the mean of the three groups.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, <span class="va">y1</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, type<span class="op">=</span><span class="st">"n"</span>, xlab<span class="op">=</span><span class="st">"Dose"</span>, ylab<span class="op">=</span><span class="st">"Response"</span><span class="op">)</span> </span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="fu">lines</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, col<span class="op">=</span><span class="fl">1</span>, lwd<span class="op">=</span><span class="fl">1</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="fu">lines</span><span class="op">(</span><span class="va">x</span>, <span class="va">z</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, col<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">1</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">x</span>, <span class="va">ym</span>, col<span class="op">=</span><span class="fl">1</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">x</span>, <span class="va">zm</span>, col<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"bottomleft"</span>, lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Treated"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/eda-avoid-plots-rendered-show-important-factors-1.png" alt="Because dose is an important factor, we show it in this plot." class="figure mx-auto d-block"><figcaption>
Because dose is an important factor, we show it in this plot.
</figcaption></figure>
</div>
<div class="section level4">
<h4 id="too-many-significant-digits">Too many significant digits<a class="anchor" aria-label="anchor" href="#too-many-significant-digits"></a>
</h4>
<p>By default, statistical software like R returns many significant
digits. This does not mean we should report them. Cutting and pasting
directly from R is a bad idea since you might end up showing a table,
such as the one below, comparing the heights of basketball players:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heights</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">8</span>,<span class="fl">73</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rnorm</span><span class="op">(</span><span class="fl">8</span>,<span class="fl">73</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rnorm</span><span class="op">(</span><span class="fl">8</span>,<span class="fl">80</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>                 <span class="fu">rnorm</span><span class="op">(</span><span class="fl">8</span>,<span class="fl">78</span>,<span class="fl">3</span><span class="op">)</span>,<span class="fu">rnorm</span><span class="op">(</span><span class="fl">8</span>,<span class="fl">78</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu">c</span><span class="op">(</span><span class="st">"SG"</span>,<span class="st">"PG"</span>,<span class="st">"C"</span>,<span class="st">"PF"</span>,<span class="st">"SF"</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">heights</span><span class="op">)</span><span class="op">&lt;-</span> <span class="fu">paste</span><span class="op">(</span><span class="st">"team"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">heights</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             SG       PG        C       PF       SF
team 1 76.39843 76.21026 81.68291 75.32815 77.18792
team 2 74.14399 71.10380 80.29749 81.58405 73.01144
team 3 71.51120 69.02173 85.80092 80.08623 72.80317
team 4 78.71579 72.80641 81.33673 76.30461 82.93404
team 5 73.42427 73.27942 79.20283 79.71137 80.30497
team 6 72.93721 71.81364 77.35770 81.69410 80.39703
team 7 68.37715 73.01345 79.10755 71.24982 77.19851
team 8 73.77538 75.59278 82.99395 75.57702 87.68162</code></pre>
</div>
<p>We are reporting precision up to 0.00001 inches. Do you know of a
tape measure with that much precision? This can be easily remedied:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">round</span><span class="op">(</span><span class="va">heights</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         SG   PG    C   PF   SF
team 1 76.4 76.2 81.7 75.3 77.2
team 2 74.1 71.1 80.3 81.6 73.0
team 3 71.5 69.0 85.8 80.1 72.8
team 4 78.7 72.8 81.3 76.3 82.9
team 5 73.4 73.3 79.2 79.7 80.3
team 6 72.9 71.8 77.4 81.7 80.4
team 7 68.4 73.0 79.1 71.2 77.2
team 8 73.8 75.6 83.0 75.6 87.7</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="displaying-data-well">Displaying data well<a class="anchor" aria-label="anchor" href="#displaying-data-well"></a>
</h4>
<p>In general, you should follow these principles:</p>
<ul>
<li>Be accurate and clear.</li>
<li>Let the data speak.</li>
<li>Show as much information as possible, taking care not to obscure the
message.</li>
<li>Science not sales: avoid unnecessary frills (especially gratuitous
3D).</li>
<li>In tables, every digit should be meaningful. Don’t drop ending
0’s.</li>
</ul>
<p>Some further reading:</p>
<ul>
<li>ER Tufte (1983) The visual display of quantitative information.
Graphics Press.</li>
<li>ER Tufte (1990) Envisioning information. Graphics Press.</li>
<li>ER Tufte (1997) Visual explanations. Graphics Press.</li>
<li>WS Cleveland (1993) Visualizing data. Hobart Press.</li>
<li>WS Cleveland (1994) The elements of graphing data. CRC Press.</li>
<li>A Gelman, C Pasarica, R Dodhia (2002) Let’s practice what we preach:
Turning tables into graphs. The American Statistician 56:121-130.</li>
<li>NB Robbins (2004) Creating more effective graphs. Wiley.</li>
<li><a href="https://bang.clearscience.info/?p=546" class="external-link">Nature Methods
columns</a></li>
</ul>
</div>
</section><section><h2 class="section-heading" id="misunderstanding-correlation-advanced">Misunderstanding Correlation (Advanced)<a class="anchor" aria-label="anchor" href="#misunderstanding-correlation-advanced"></a>
</h2>
<hr class="half-width">
<p>The use of correlation to summarize reproducibility has become
widespread in, for example, genomics. Despite its English language
definition, mathematically, correlation is not necessarily informative
with regards to reproducibility.<br>
Here we briefly describe three major problems.</p>
<p>The most egregious related mistake is to compute correlations of data
that are not approximated by bivariate normal data. As described above,
averages, standard deviations and correlations are popular summary
statistics for two-dimensional data because, for the bivariate normal
distribution, these five parameters fully describe the distribution.
However, there are many examples of data that are not well approximated
by bivariate normal data. Raw gene expression data, for example, tends
to have a distribution with a very fat right tail.</p>
<p>The standard way to quantify reproducibility between two sets of
replicated measurements, say <span class="math inline">\(x_1,\dots,x_n\)</span> and <span class="math inline">\(y_1,\dots,y_n\)</span>, is simply to compute the
distance between them:</p>
<p><span class="math display">\[
\sqrt{
\sum_{i=1}^n d_i^2} \mbox{ with } d_i=x_i - y_i
\]</span></p>
<p>This metric decreases as reproducibility improves and it is 0 when
the reproducibility is perfect. Another advantage of this metric is that
if we divide the sum by N, we can interpret the resulting quantity as
the standard deviation of the <span class="math inline">\(d_1,\dots,d_N\)</span> if we assume the <span class="math inline">\(d\)</span> average out to 0. If the <span class="math inline">\(d\)</span> can be considered residuals, then this
quantity is equivalent to the root mean squared error (RMSE), a summary
statistic that has been around for over a century. Furthermore, this
quantity will have the same units as our measurements resulting in a
more interpretable metric.</p>
<p>Another limitation of the correlation is that it does not detect
cases that are not reproducible due to average changes. The distance
metric does detect these differences. We can rewrite:</p>
<p><span class="math display">\[\frac{1}{n} \sum_{i=1}^n (x_i - y_i)^2 =
\frac{1}{n} \sum_{i=1}^n [(x_i - \mu_x) - (y_i - \mu_y) + (\mu_x -
\mu_y)]^2\]</span></p>
<p>with <span class="math inline">\(\mu_x\)</span> and <span class="math inline">\(\mu_y\)</span> the average of each list. Then we
have:</p>
<p><span class="math display">\[\frac{1}{n} \sum_{i=1}^n (x_i - y_i)^2 =
\frac{1}{n} \sum_{i=1}^n (x_i-\mu_x)^2 +  \frac{1}{n} \sum_{i=1}^n (y_i
- \mu_y)^2 + (\mu_x-\mu_y)^2 + \frac{1}{n}\sum_{i=1}^n (x_i-\mu_x)(y_i -
\mu_y)
\]</span></p>
<p>For simplicity, if we assume that the variance of both lists is 1,
then this reduces to:</p>
<p><span class="math display">\[\frac{1}{n} \sum_{i=1}^n (x_i - y_i)^2 =
2 + (\mu_x-\mu_y)^2 - 2\rho\]</span></p>
<p>with <span class="math inline">\(\rho\)</span> the correlation. So we
see the direct relationship between distance and correlation. However,
an important difference is that the distance contains the term <span class="math display">\[(\mu_x-\mu_y)^2\]</span> and, therefore, it can
detect cases that are not reproducible due to large average changes.</p>
<p>Yet another reason correlation is not an optimal metric for
reproducibility is the lack of units. To see this, we use a formula that
relates the correlation of a variable with that variable, plus what is
interpreted here as deviation: <span class="math inline">\(x\)</span>
and <span class="math inline">\(y=x+d\)</span>. The larger the variance
of <span class="math inline">\(d\)</span>, the less <span class="math inline">\(x+d\)</span> reproduces <span class="math inline">\(x\)</span>. Here the distance metric would depend
only on the variance of <span class="math inline">\(d\)</span> and would
summarize reproducibility. However, correlation depends on the variance
of <span class="math inline">\(x\)</span> as well. If <span class="math inline">\(d\)</span> is independent of <span class="math inline">\(x\)</span>, then</p>
<p><span class="math display">\[
\mbox{cor}(x,y) = \frac{1}{\sqrt{1+\mbox{var}(d)/\mbox{var}(x)}}
\]</span></p>
<p>This suggests that correlations near 1 do not necessarily imply
reproducibility. Specifically, irrespective of the variance of <span class="math inline">\(d\)</span>, we can make the correlation
arbitrarily close to 1 by increasing the variance of <span class="math inline">\(x\)</span>.</p>
<p>When is it appropriate to use pie charts or donut charts?<br>
•A) When you are hungry.<br>
•B) To compare percentages.<br>
•C) To compare values that add up to 100%.<br>
•D) Never.</p>
<p>The use of pseudo-3D plots in the literature mostly adds: •A)
Pizzazz.<br>
•B) The ability to see three dimensional data.<br>
•C) Ability to discover.<br>
•D) Confusion.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Pie charts, donut charts, and 3-dimensional plots should not be used
to display data.</li>
<li>Box plots are a better alternative to bar plots.</li>
<li>If data are small, data points can be shown in the plot.</li>
</ul>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/statistical-inference-for-biology/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/statistical-inference-for-biology/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, statistics, inference, biology",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/statistical-inference-for-biology/aio.html",
  "identifier": "https://carpentries-incubator.github.io/statistical-inference-for-biology/aio.html",
  "dateCreated": "2020-02-25",
  "dateModified": "2025-08-19",
  "datePublished": "2025-08-19"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

